---
title: "R Notebook of e12r1 downstream analysis from binarized peak level data with joint feature space"
output: html_notebook
---

```{r Load libraries, include=FALSE}
if (Sys.info()['sysname']=="Linux") {
    .libPaths(c("/projappl/project_2001539/project_rpackages", .libPaths()))
    libpath <- .libPaths()[1]
}

load.libs <- c(
  "viridisLite",
  "DT",
  "GenomicRanges",
  "data.table",
  "plyr",
  "gplots",
  "Matrix",
  "proxy",
  "Rtsne",
  "genomation",
  "densityClust",
  "irlba",
  "umap",
  "clusterProfiler",
  "RColorBrewer",
  "fastcluster",
  "gridExtra",
  "Signac",
  "Seurat",
  "regioneR",
  "reldist",
  "SeuratWrappers",
  "chromVAR",
  "tidyverse",
  "TFBSTools",
  "BSgenome.Mmusculus.UCSC.mm10",
  "EnsDb.Mmusculus.v79",
  "org.Mm.eg.db",
  "pheatmap",
  "plotly",
  "waffle",
  "ggplotify",
  "flexdashboard",
  "motifmatchr",
  "future",
  "cicero",
  "universalmotif")
if (!require("pacman")) install.packages("pacman"); library(pacman)
p_load(load.libs, update = FALSE, character.only = TRUE)
status <- sapply(load.libs,require,character.only = TRUE)
if(all(status)){
    print("SUCCESS: You have successfully installed and loaded all required libraries.")
} else{
    cat("ERROR: One or more libraries failed to install correctly. Check the following list for FALSE cases and try again...\n\n")
    status
}

set.seed(2020);
```
About 
===================================== 
Process from counts matrix to ready for downstream analysis

In this phase we do final filtering, binarization, TF-IDF, SVD, UMAP for scATAC data based on binarized count over peaks identified over clades in phases 1-3, extract transcript expression signal from scATAC data, plot marker genes, find tentative cluster ids, integrate with scRNA etc...


```{r Aux functions and number of cores, include=FALSE}
if (Sys.info()['sysname']=="Darwin") {
    source("../../generic code/AuxFunctions.R")
    cores=3
} else {
    source("~/Research/⁨GitHub/⁨scATAC-pipelines⁩/AuxFunctions.R")
    cores=10
}
```

```{r Setting Seurat multicore}
plan("multisession", workers = 6)
options(future.globals.maxSize = 12 * 1024 ^ 3)
```

```{r Setwd sample name specific variables and read data objects, message=FALSE}
#setwd("/Volumes/ExtSSD/ExtraWorkspace/E12R1/scATAC_data/")
sample.name <- "E12_R1"
run.date <- "240322"
# Reading scRNA data rds object
#s.data_rna <- readRDS("../scRNA/E14_DI_scRNAseq_neurons_clean.rds")
#cluster_names<-read_tsv("../scRNA/E14_DI_scRNAseq_cleaned_top20_allclusters.csv")
#cluster_id_name <- distinct(cluster_names[,c("cluster","ClusterName")])->cluster_id_name
#scRNA_clean_markers_file <- "../scRNA/E14_DI_scRNAseq_cleaned_top20_allclusters.csv"
```

```{r Reading in data objects}
s.data <- readRDS(paste("../scATAC_data/",sample.name,".merged.peaks.271021.Rds",sep=""))
s.data_RNA <- readRDS("../scRNA_data/e12_ens_whole_rescaled.Rds")
s.data_RNA$clusterAnnotation <- s.data_RNA$seurat_clusters
feature.metadata.rna <- s.data_RNA[['RNA']][[]] %>% rownames_to_column("gene_id") %>% as_tibble()
```


```{r TFIDF and SVD, cache=TRUE, message=FALSE}
DefaultAssay(s.data) <- 'peaks'

set.seed(2020)

s.data <- RunTFIDF(s.data, method=3)
s.data <- FindTopFeatures(s.data, min.cutoff = 'q25')
s.data <- RunSVD(
  object = s.data,
  assay = 'peaks',
  reduction.key = 'LSI_',
  reduction.name = 'lsi'
)
```


Visualizations {data-icon="fa-signal"}
=====================================

### Read depth and dimension reduction component correlation

```{r plot_depth_dimension_corr}
DepthCor(s.data)
```

---

Plotting read depth correlation plot, in Seurat pipeline usually first PC1 is omitted because of this association.

### scATAC UMAP clusters

```{r UMAP and cluster detection, cache=TRUE}
s.data <- RunUMAP(object = s.data, reduction = 'lsi', dims = 2:30, min.dist=0.05, spread=1.1)
s.data <- FindNeighbors(object = s.data, reduction = 'lsi', dims = 2:30)
s.data <- FindClusters(object = s.data, verbose = FALSE, algorithm = 4, resolution = 1)
DimPlot(object = s.data, label = TRUE, pt.size=1.2) + NoLegend()
```

```{r UMAP with biological replicates colored}
DimPlot(object = s.data, label = TRUE, pt.size=.5,group.by="replicate", shuffle =TRUE) + NoLegend()
```

```{r Create a gene activity matrix}
# Compute gene activities
# Now that scRNA data has been processed with ENSMUSG ids this now longer works directly like this.
# Let's extract genomic annotation metadata from s.data[['peaks']]
genomic.metadata <- mcols(Annotation(s.data[['peaks']]))
# Generate conversion table from gene_name to gene_id
gene_name2gene_id <- as_tibble(genomic.metadata[,c("gene_name","gene_id")])

# Calculate gene activity estimate from scATAC reads based on the scATAC, by using gene_names as function does not support any other id
gene.activities <- GeneActivity(s.data, assay="peaks")

# Store gene_names
gene.names <- rownames(gene.activities)

# Switch sparse matrix to use ensmusg id
ensmusg.ids <- gene_name2gene_id[match(gene.names,pull(gene_name2gene_id,"gene_name")),] %>% pull("gene_id")
gene_names <- gene_name2gene_id[match(gene.names,pull(gene_name2gene_id,"gene_name")),] %>% pull("gene_name")

# Dropping NAs
non.na.i <- !is.na(ensmusg.ids)
gene.activities.gene_id <- gene.activities[non.na.i,]
rownames(gene.activities.gene_id) <- ensmusg.ids[non.na.i]

# Add the gene activity matrix to the Seurat object as a new assay
s.data[['Activity']] <- CreateAssayObject(counts = gene.activities.gene_id)
s.data <- NormalizeData(
  object = s.data,
  assay = 'Activity',
  normalization.method = 'LogNormalize',
  scale.factor = median(s.data$nCount_Activity)
)

# Add gene_name to gene_id mapping into the s.data[['Activity']] assays metadata
s.data[['Activity']]<- AddMetaData(s.data[['Activity']], col.name = "feature_symbol", metadata = gene_names[non.na.i])
```

### scATAC clades projection into UMAP clusters

```{r Testing how clades from phase 1 situate in UMAP projection}
#DimPlot(object = s.data, label = TRUE, pt.size=1.2,group.by="clade") + NoLegend()
```

### Read marker gene list and do FeaturePlot

```{r Read marker gene list and do FeaturePlot, fig.height=20, fig.width=20, message=FALSE}
neuronal.markers<- read_tsv("../../CellAnnotation/E12.5_cluster_markers_for_ATACseq.txt", col_names = c("annotation","geneName"))
feature.metadata <- s.data[['Activity']][[]] %>% rownames_to_column(var="gene_id") %>% as_tibble()
neuronal.markers.tmp <- filter(feature.metadata, feature_symbol %in% neuronal.markers$geneName)

neuronal.markers.ids <- pull(neuronal.markers.tmp,"gene_id")
neuronal.markers.names <- pull(neuronal.markers.tmp,"feature_symbol")

DefaultAssay(s.data) <- 'Activity'

f.plot.tmp <- FeaturePlot(
  object = s.data,
  features = neuronal.markers.ids,
  pt.size = 0.1,
  max.cutoff = 'q95',
  combine = F
)

f.plots.1 <- lapply(1:length(f.plot.tmp),function(i){
  f.plot.tmp[[i]] + labs(title=neuronal.markers.names[i])
})

patchwork::wrap_plots(f.plots.1)
```

----

Plot of some marker genes with signal from 2kbp upstream and only from those feature 
regions (peaks) that form clusters seen in UMAP


### Perform integration with RNA sample

```{r Adding E12R1 cluster information from Morello et al.}
# morello_e12_clusters <- read_tsv("../scRNA_data/Morello_et_al_cluster_anno.tsv")
# 
# morello.i <- match(as.vector(s.data_RNA@meta.data$seurat_clusters),morello_e12_clusters$clusterNumber)
# 
# s.data_RNA <- AddMetaData(s.data_RNA, pull(morello_e12_clusters[morello.i,2], clusterName), col.name = "clusterAnnotation")
```

```{r Perform label transfer as per Satija et al, cache=TRUE}
# Converting numerical Seurat clusters to alphabeticals and store in $CellType slot of s.data_rna

# Finding transfer anchors
transfer.anchors <- FindTransferAnchors(
    reference = s.data_RNA,
    query = s.data,
    reduction = 'cca',
    reference.assay="RNA",
    query.assay = "Activity",
    features = VariableFeatures(object=s.data_RNA)
)

# TODO: Find out which object and why there are graphs(4?) without associated assays, this however based on googling doesn't show as meaningful warning()

predicted.labels <- TransferData(
  anchorset = transfer.anchors,
  refdata = s.data_RNA@meta.data$clusterAnnotation,
  weight.reduction = s.data[['lsi']],
  dims = 2:30
)

s.data <- AddMetaData(object = s.data, metadata = predicted.labels)
```

----

Label transfer, following and adapting https://satijalab.org/seurat/v3.0/atacseq_integration_vignette.html

### Plotting prediction scores

```{r Plot prediction scores, echo=FALSE}
pred.score.df <- data.frame(pred.score=s.data$prediction.score.max)
ggplot(pred.score.df, aes(x=pred.score)) + geom_histogram(binwidth=.025) + geom_vline(data=pred.score.df, aes(xintercept=0.5, color="red"),linetype="dashed") + theme(legend.position = 'none')
```

---

Histogram of prediction scores, in Satija lab integration process each cell get prediction score how well it can be mapped from dataspace to another. Values > 0.5 are considered to be "good".

### Fraction of cells mapping acceptable between scATAC and scRNA data spaces

```{r Calculate accepted prediction score fraction}
prediction.score.over.th <- table(s.data$prediction.score.max > 0.5)
p.freq <- prediction.score.over.th['TRUE']/(prediction.score.over.th['TRUE']+prediction.score.over.th['FALSE'])

p.score.th <- as.numeric(prediction.score.over.th)
val_names <- sprintf("%s (%s)", c("Match not found", "Match found"), scales::percent(round(p.score.th/sum(p.score.th), 2)))
names(p.score.th) <- val_names
waffle::waffle(p.score.th/10,colors = c("#fb8072", "#8dd3c7", "white"), rows=9, size=1)
```

---

Waffle plot visualizing proportions of properly mapping cells

### Label transfer visualization

```{r Select cells with accepted prediction score and plot results of label transfer, fig.height=10, fig.width=18, cache=TRUE}
#' Select only cells with prediction score over 0.5
s.data.filtered <- subset(s.data, subset = prediction.score.max > 0.5)

#' To make the colors match, TODO: Check why there are few NAs in the predicted ids
#s.data.filtered$predicted.id <- factor(s.data.filtered$predicted.id, levels = letters[as.numeric(levels(s.data_rna))]) 

# Do combined plotting
p1 <- DimPlot(s.data.filtered, group.by = "predicted.id", label = TRUE, repel = TRUE, label.size=7, pt.size=2) + NoLegend() + scale_colour_hue(drop = FALSE)
p2 <- DimPlot(s.data_RNA, group.by = "clusterAnnotation", label = TRUE, repel = TRUE, label.size=7) + NoLegend()

p1 + ggtitle("scATAC-seq cells, labels predicted from scRNA")
p2 + ggtitle("scRNA-seq cells")
```

---

Filtering for cells mapping properly and visualizing cluster labels from scRNA (right side) to scATAC (left side).

### Perform RNA data imputation into scATAC cells

```{r Perform scRNA data imputation, message=FALSE, include = FALSE}
#genes.use <- VariableFeatures(s.data_rna)
refdata <- GetAssayData(s.data_RNA, assay = "RNA", slot = "data")#[genes.use, ]

s.data_RNA@meta.data$tech<-"scRNA"
s.data@meta.data$tech<-"scATAC"

# refdata (input) contains a scRNA-seq expression matrix for the scRNA-seq cells.  
#' Imputation (output) will contain an imputed scRNA-seq matrix for each of the ATAC cells
imputation <- TransferData(anchorset = transfer.anchors, refdata = refdata, weight.reduction = s.data[["lsi"]], dims = 2:30)

# this line adds the imputed data matrix to the pbmc.atac object
s.data[["RNA"]] <- imputation
coembed <- merge(x = s.data_RNA, y = s.data)

# Copy feature metadata from s.data_rna to s.data
s.data_rna.feature.metadata <- s.data_RNA[["RNA"]][[]]
s.data[["RNA"]] <- AddMetaData(s.data[["RNA"]], metadata = s.data_rna.feature.metadata[rownames(s.data[["RNA"]]),"feature_symbol"], col.name = "feature_symbol")

# Find variable features
coembed <- FindVariableFeatures(coembed)

# Finally, we run PCA and UMAP on this combined object, to visualize the co-embedding of both
# datasets
coembed <- ScaleData(coembed, do.scale = FALSE)
coembed <- RunPCA(coembed, verbose = FALSE)
coembed <- RunUMAP(coembed, dims = 2:30)
coembed@meta.data$clusterAnnotation <- ifelse(!is.na(coembed@meta.data$clusterAnnotation), coembed@meta.data$clusterAnnotation, coembed@meta.data$predicted.id)
```

    Finding integration vectors
    
    Finding integration vector weights
    
    Transfering 27999 features onto reference data
    
    Centering data matrix
    
    11:43:23 UMAP embedding parameters a = 0.9922 b = 1.112
    
    11:43:23 Read 10354 rows and found 29 numeric columns
    
    11:43:23 Using Annoy for neighbor search, n_neighbors = 30
    
    11:43:23 Building Annoy index with metric = cosine, n_trees = 50
    
    0%   10   20   30   40   50   60   70   80   90   100%
    
    [----|----|----|----|----|----|----|----|----|----|
    
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    *
    
    |
    
    11:43:28 Writing NN index file to temp file /var/folders/yb/wvk8t07s719136klxnc8nmp9sglf93/T//Rtmpnq8Zdj/file72653e114263
    
    11:43:28 Searching Annoy index using 1 thread, search_k = 3000
    
    11:43:32 Annoy recall = 100%
    
    11:43:32 Commencing smooth kNN distance calibration using 1 thread
    
    11:43:35 Initializing from normalized Laplacian + noise
    
    11:43:35 Commencing optimization for 200 epochs, with 432826 positive edges
    
    11:43:45 Optimization finished
    



### Coembed dataset plotting

```{r Plot coembed data based on technology and cell type, cache=TRUE}
p1 <- DimPlot(coembed, group.by = "tech")
p2 <- DimPlot(coembed, group.by = "clusterAnnotation", label = TRUE, repel = TRUE)  + theme(legend.position="none") 

p1 + p2
```

---

Coembed plotting scATAC and scRNA cells together. Usable mainly for validation purposes.

```{r Filtering Seurat data object based on prediction score for all downstream analysis}
s.data <- subset(s.data, subset = prediction.score.max > 0.5)
```

```{r Regenerate UMAP after filtering based predictions score}
DefaultAssay(s.data) <- "peaks"
s.data <- RunUMAP(object = s.data, reduction = 'lsi', dims = 2:30, spread=1.4)
s.data <- FindNeighbors(object = s.data, reduction = 'lsi', dims = 2:30)
s.data <- FindClusters(object = s.data, verbose = FALSE, algorithm = 4, resolution = 1)
DimPlot(object = s.data, label = TRUE, pt.size=1.2) + NoLegend()
```


### Plot marker gene FeaturePlots with imputed scRNA data

```{r Plot marker gene FeaturePlots with imputed scRNA data, fig.height=20, fig.width=20, cache=TRUE, message=FALSE}
neuronal.markers<- read_tsv("../../CellAnnotation/E12.5_cluster_markers_for_ATACseq.txt", col_names = c("annotation","geneName"))
feature.metadata <- s.data[['Activity']][[]] %>% rownames_to_column(var="gene_id") %>% as_tibble()
neuronal.markers.tmp <- filter(feature.metadata, feature_symbol %in% neuronal.markers$geneName)

neuronal.markers.ids <- pull(neuronal.markers.tmp,"gene_id")
neuronal.markers.names <- pull(neuronal.markers.tmp,"feature_symbol")
DefaultAssay(s.data) <- 'RNA'

f.plot.tmp <- FeaturePlot(
  object = s.data,
  features = neuronal.markers.ids,
  pt.size = 0.1,
  max.cutoff = 'q95',
  combine = F
)

f.plots.2 <- lapply(1:length(f.plot.tmp),function(i){
  f.plot.tmp[[i]] + labs(title=neuronal.markers.names[i])
})

patchwork::wrap_plots(f.plots.2)
```

---

Plot marker gene FeaturePlots with imputed scRNA data

### Plot gini index based validation of clustering effectiveness

```{r Plot gini index based validation of clustering effectiveness, message=FALSE}
# Define a set of HK genes
hk.genes <- c("RRN18S","Actb","Gapdh","Pgk1","Ppia","Rpl13a","Rplp0","Arbp","B2M","Ywhaz","Sdha","Tfrc","Gusb","Hmbs","Hprt1","Tbp")

hk.genes.id <- convert_feature_identity(s.data, "RNA",features = hk.genes)
neuronal.markers.id <- convert_feature_identity(s.data, "RNA",features = neuronal.markers$geneName)

# Neuronal marker mean per cluster in Cusanovich data
gene.i <- match(neuronal.markers.id,s.data[['RNA']]@data@Dimnames[[1]])
gene.i<-gene.i[!is.na(gene.i)]
barcode.clusters <- s.data@meta.data$seurat_clusters
marker.matrix <- s.data[['RNA']]@data[gene.i,]
marker.tb <- as_tibble(t(as.data.frame(marker.matrix)))
marker.tb<-tibble(marker.tb,cluster=barcode.clusters)
marker.mean <- list()
marker.mean$mean.by.cluster <- marker.tb %>% group_by(cluster) %>% summarize_all(mean)

# HK gene mean per cluster in Cusanovich data
gene.i <- match(hk.genes.id,s.data[['RNA']]@data@Dimnames[[1]])
gene.i<-gene.i[!is.na(gene.i)]
barcode.clusters <- s.data@meta.data$seurat_clusters
marker.matrix <- s.data[['RNA']]@data[gene.i,]
marker.tb <- as_tibble(t(as.data.frame(marker.matrix)))
marker.tb<-tibble(marker.tb,cluster=barcode.clusters)
hk.mean <- list()
hk.mean$mean.by.cluster <- marker.tb %>% group_by(cluster) %>% summarize_all(mean)

# Gini indeces for Cusanovich data
cus.hk.gini <- apply(hk.mean$mean.by.cluster[,-1],2,gini)
cus.neur.gini <- apply(marker.mean$mean.by.cluster[,-1],2,gini)

gini.tb<-tibble(gini.index=c(cus.hk.gini,cus.neur.gini),type=c(rep("hk",length(cus.hk.gini)),rep("neur",length(cus.neur.gini)))) %>%  dplyr::filter(!is.na(gini.index))

ggplot(gini.tb, aes(x=gini.index,y=type,fill="blue"))+geom_boxplot(fill="lightblue")+ theme(legend.position="none") + theme_classic() 
```

---

Gini index based validation of clustering. 

### Adding Motif information into the object

```{r Adding Motif information to the object}
Hocomocov11 <- read_jaspar("../../mm10/HOCOMOCOv11_core_MOUSE_mono_jaspar_format.txt")
names(Hocomocov11) <- lapply(Hocomocov11,function(x){x@name})
Hocomocov11 <- convert_motifs(Hocomocov11, "TFBSTools-PWMatrix")
PWMs <- do.call(PWMatrixList,Hocomocov11)

DefaultAssay(s.data) <- "peaks"

# add motif information
s.data <- Signac::AddMotifs(
  object = s.data,
  genome = BSgenome.Mmusculus.UCSC.mm10,
  pfm = PWMs
)
```

```{r Finding closest features}
DefaultAssay(s.data) <- "peaks"
closest.features <- ClosestFeature(s.data, regions = rownames(s.data))
saveRDS(closest.features, file="../analyses/E12R1_nmm_closest_features.271021.Rds")
```

```{r Run ChromVar}
s.data <- RunChromVAR(
  object = s.data,
  genome = BSgenome.Mmusculus.UCSC.mm10
)
```

### Identification of markers for clusters defined based on both modalities

```{r Identification of cluster markers, message=FALSE}
# We need to run detection separately for both modality and then combine via AUC score
  
print("Running presto::wilcoxauc for RNA modality")
DefaultAssay(s.data) <- "RNA"
markers_rna <- presto:::wilcoxauc.Seurat(X = s.data, group_by = "seurat_clusters", assay = 'data', seurat_assay = 'RNA')

print("Running presto::wilcoxauc for ATAC modality")
DefaultAssay(s.data) <- "peaks"
markers_atac <- presto:::wilcoxauc.Seurat(X = s.data, group_by = "seurat_clusters", assay = 'data', seurat_assay = 'peaks')

markers.atac.annotated <- as_tibble(cbind(markers_atac, closest.features))
# Then we need to 1) annotate ATAC features 2) combine with RNA modality 3) Think of its presentation

#saveRDS(atac.expression.markers, file = paste("../analyses/", sample.name,".atac.expression.markers.mm.Rds",sep=""))
```

### Finding overrepresented motifs among markers_atac features

```{r Finding overrepresented motifs among markers_atac features, message=FALSE}
motif.markers <- markers.atac.annotated %>% filter(logFC > 0.25 & padj <= 0.01) %>% group_by(group) %>% select(feature, group) %>% group_modify(~FindMotifs(object=s.data, features=.x$feature)) %>% filter(pvalue <= 0.01 & fold.enrichment >= 1.5)
```

```{r Find markers for clusters based on chromvar, message=FALSE}
DefaultAssay(s.data) <- "chromvar"
markers_chromvar <- as_tibble(FindAllMarkers(
  object = s.data,
  only.pos = TRUE,
  test.use = 'LR',
  latent.vars = 'nCount_peaks'
)) %>% filter(p_val_adj <= 0.01 & avg_log2FC >= 0.75)
```

```{r Filter marker data to be saved}
# Now we need to combine markers_rna, markers_atac, markers_chromvar, motif.markers in meaningful output to help with cluster annotation

# Writing marker info out to be used separately, writing out top 100
top.markers_rna <- as_tibble(markers_rna) %>% dplyr::filter(padj <= 0.01) %>% group_by(group) %>% top_n(n = 25, wt = logFC)
top.markers_rna <- left_join(top.markers_rna, feature.metadata.rna, by=c("feature"="gene_id"))

top.markers_atac <- as_tibble(markers.atac.annotated) %>% dplyr::filter(padj <= 0.01) %>% group_by(group) %>% top_n(n = 25, wt = logFC)

top.markers_chromvar <- as_tibble(markers_chromvar) %>% dplyr::filter(p_val_adj <= 0.01) %>% group_by(cluster) %>% top_n(n = 25, wt = avg_log2FC)

top.markers_motifs <- as_tibble(motif.markers) %>% dplyr::filter(pvalue <= 0.01) %>% group_by(group) %>% top_n(n = 25, wt = fold.enrichment)

save(list=c("top.markers_rna","top.markers_atac","top.markers_chromvar","top.markers_motifs"), file=paste("../analyses/e12R1_nmm_scATAC_cluster_markers.",run.date,".RData",sep=""))
```

### Saving data for downstream analyses

```{r Saving data for downstream analyses}
saveRDS(s.data,paste("../scATAC_data/",sample.name,"_DownstreamReady_nmm_.",run.date,".Rds",sep=""))
```

```{r Saving data for downstream analyses slim}
s.data.slim <- s.data
DefaultAssay(s.data.slim) <- "peaks"
s.data.slim[['peaks_count']] <- NULL
s.data.slim[['Activity']] <- NULL
saveRDS(s.data.slim,paste("../scATAC_data/",sample.name,"_DownstreamReady_nmm_slim.",run.date,".Rds",sep=""))
```

```{r}
pdf(file="/Users/kilpinen/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Figures/plots/E12.5.scATAC.UMAP.full.pdf", width = 8, height = 8)
DimPlot(s.data) + coord_fixed()
dev.off()
```


```{r}
sessionInfo()
```


Tables {data-icon="fa-table"}
=====================================

### Cross-tabulation (confusion matrix) in to what extent each scRNA based cell type is included in each scATAC cluster.

```{r Cross tabulation between scATAC and scRNA based clustering}
scATAC.clusters <- Idents(s.data)
scRNA.clusters <- s.data@meta.data$predicted.id

conf.mat <- table(as.factor(scRNA.clusters),scATAC.clusters)
create_dt(as.data.frame.matrix(conf.mat))
```

---

Cross-tabulation (confusion matrix) in to what extent each scRNA based cell type is included in each scATAC cluster.


### Table format of the top1 gene expression by scATAC cluster dotplot

```{r, message=FALSE}
# create_dt(as.data.frame.matrix(d.plot$data))
```

---

### Top 20 marker genes per scATAC cluster

```{r Top 20 marker genes per scATAC cluster}
# top.20.markers.by.cluster <- e14di.atc.expression.markers %>% group_by(cluster) %>% top_n(n = 20, wt = avg_log2FC)
# create_dt(as.data.frame.matrix(data.frame(top.20.markers.by.cluster)))
```

---

Marker table filtered for convenience having top 20 genes pre cluster

### scRNA cluster marker genes (top20 per cluster)

```{r scRNA cluster marker genes, message=FALSE}
# clean_top20_markers <- read_tsv(scRNA_clean_markers_file)
# create_dt(clean_top20_markers)
```

---
