---
title: "R Notebook"
output: html_notebook
---

```{r}
library(tidyverse)
library(DBI)
library(readxl)
library(writexl)
library(GenomicRanges)
```

```{r}
con.obj <- DBI::dbConnect(RSQLite::SQLite(), dbname = "~/Workspace/TOBIAS.dr.h12_300824.sqlite")
table.loc <- "/Users/kilpinen/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Tables/Per gene feature tables/"
```

## Gata2

```{r}
Gata2.features <- readxl::read_xlsx("/Users/kilpinen/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Tables/Archive/Gata2 feature summary_new.xlsx", sheet = "Sheet1")
```

```{r}
links.tb <- as_tibble(dbGetQuery(con.obj, 'SELECT li.*,gm.gene_name FROM links as li, gene_metadata as gm WHERE gm.ensg_id=li.ensg_id'))
```

```{r}
Gata2.features <- distinct(left_join(Gata2.features, distinct(dplyr::filter(links.tb, gene_name=="Gata2")), by=c("Feature"="feature")))
```

```{r}
Gata2.tb.out <- Gata2.features %>% dplyr::select("Gene_Locus","Feature_number","Feature","kb from TSS","length bp","score","zscore", "pvalue","Gene_identity","Comment","Ref")
writexl::write_xlsx(Gata2.tb.out, path=paste(table.loc,"Gata2_features.xlsx",sep=""))
```

## Gata3

```{r}
Gata3.features <- readxl::read_xlsx("/Users/kilpinen/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Tables/Archive/Gata3 feature summary_new.xlsx", sheet = "Sheet1")
```

```{r}
links.tb <- as_tibble(dbGetQuery(con.obj, 'SELECT li.*,gm.gene_name FROM links as li, gene_metadata as gm WHERE gm.ensg_id=li.ensg_id'))
```

```{r}
Gata3.features <- distinct(left_join(Gata3.features, distinct(dplyr::filter(links.tb, gene_name=="Gata3")), by=c("Feature"="feature")))
```

```{r}
Gata3.tb.out <- Gata3.features %>% dplyr::select("Gene_Locus","Feature_number","Feature","kb from TSS","length bp","score","zscore", "pvalue","Gene_identity")
writexl::write_xlsx(Gata3.tb.out, path=paste(table.loc,"Gata3_features.xlsx",sep=""))
```

## Tal1

```{r}
Tal1.features <- readxl::read_xlsx("/Users/kilpinen/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Tables/Archive/Tal1 feature summary_new_2.xlsx", sheet = "Sheet1")
```

```{r}
links.tb <- as_tibble(dbGetQuery(con.obj, 'SELECT li.*,gm.gene_name FROM links as li, gene_metadata as gm WHERE gm.ensg_id=li.ensg_id'))
```

```{r}
Tal1.features <- distinct(left_join(Tal1.features, distinct(dplyr::filter(links.tb, gene_name=="Tal1")), by=c("Feature"="feature")))
```

```{r}
Tal1.tb.out <- Tal1.features %>% dplyr::select("Gene_Locus","Feature_number","Feature","kb from TSS","length bp","score","zscore", "pvalue","Gene_identity")
writexl::write_xlsx(Tal1.tb.out, path=paste(table.loc,"Tal1_features.xlsx",sep=""))
```


## Vsx2

```{r}
Vsx2.features <- readxl::read_xlsx("/Users/kilpinen/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Tables/Archive/Vsx2 feature summary_new.xlsx", sheet = "Sheet1")
```

```{r}
links.tb <- as_tibble(dbGetQuery(con.obj, 'SELECT li.*,gm.gene_name FROM links as li, gene_metadata as gm WHERE gm.ensg_id=li.ensg_id'))
```

```{r}
Vsx2.features <- distinct(left_join(Vsx2.features, distinct(dplyr::filter(links.tb, gene_name=="Vsx2")), by=c("Feature"="feature")))
```

```{r}
Vsx2.tb.out <- Vsx2.features %>% dplyr::select("Gene_Locus","Feature_number","Feature","kb from TSS","length bp","score","zscore", "pvalue","Gene_identity","Comment","Ref")
writexl::write_xlsx(Vsx2.tb.out, path=paste(table.loc,"Vsx2_features.xlsx",sep=""))
```

```{r}
# List of tibbles (replace with actual tibbles)
tibble_list <- list(Vsx2.features, Tal1.features, Gata3.features, Gata2.features)


# Function to convert each tibble to GRanges
tibble_to_grange <- function(tibble, gene_name) {
  # Ensure sequence levels are unique and non-NA
  seqlevels <- unique(na.omit(unlist(lapply(tibble_list, function(x) x$Chromosome))))
  
  tibble %>%
    # Filter rows with valid coordinates to avoid NA entries
    filter(!is.na(Chromosome) & !is.na(start.x) & !is.na(end.x)) %>%
    # Create GRanges object manually
    { 
      gr <- GRanges(
        seqnames = .$Chromosome,
        ranges = IRanges(start = .$start.x, end = .$end.x),
        Feature = .$Feature,  # Assuming Feature is useful in metadata
        gene_name = gene_name  # Add gene name as metadata
      )
      # Set consistent sequence levels without NA or duplicates
      seqlevels(gr, pruning.mode = "coarse") <- seqlevels
      gr
    }
}

# Apply the function to each tibble in the list with corresponding gene names
gene_names <- c("Vsx2", "GeneX", "GeneY", "GeneZ") # Adjust based on actual gene names
granges_list <- mapply(tibble_to_grange, tibble_list, gene_names, SIMPLIFY = FALSE)

# Combine all GRanges into one
combined_granges <- do.call(c, granges_list)

# Inspect the resulting GRanges object
combined_granges
```

```{r}
saveRDS(combined_granges,file="../analysis/Selector.features.Rds")
```

