---
title: "R Notebook of Figure 3"
output: html_notebook
---

```{r Packages, message=FALSE}
library(Signac)
library(Seurat)
library(dplyr)
library(tidyverse)
library(RColorBrewer)
library(ComplexHeatmap)
library(qs)
library(circlize)
library(GenomicRanges)
library(patchwork)
library(factoextra)
library(ggnewscale)
library(viridis)
library(dbplyr)
library(DBI)
library(clipr)
#library(igvR)
source("../../AuxCode/TOBIAS_tools.R")
source("../../AuxCode/AuxFunctions.R")
```

```{r Connection to SQLite}
con <- DBI::dbConnect(RSQLite::SQLite(), dbname = "~/Workspace/TOBIAS.dr.h12.sqlite")

tobias.table <- tbl(con, "tobias")
exp.table <- tbl(con, "exp")
acc.table <- tbl(con, "acc")
```

```{r}
plot_loc <- "~/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Figures/plots/Fig3/"
```

```{r Setting some areas of interest}
tal1.feat6 <- "chr4-115056171-115058012"
tal1.feat11 <- "chr4-115078530-115080241"
tal1.feat14 <- "chr4-115095346-115097412"
gata2.feat14 <- "chr6-88193309-88194492"
vsx2.feat5 <- "chr12-84520186-84521053"
vsx2.feat12 <- "chr12-84547959-84548872"
vsx2.feat15 <- "chr12-84569186-84572445"
vsx2.feat17 <- "chr12-84589595-84590947"
```

```{r Generating bunch of bedgraphs for IGV}
# writeBED.SQL.data(fe2, "Tal1f6_cons", "mean_cons")
# writeBED.SQL.data(fe2, filename="Tal1f6_cons_TF_motif", meta_column = "TFBS_name")
# 
# cons.filt <- qread(file="../metadata/cons.filt.gr.qs", nthreads = 8)
```

```{r Tal1f6}
Tal1f6.p1 <- plotHorizDotplot(feature.coords = tal1.feat6)

pdf(file=paste(plot_loc,"Tal1f6_dotplot.pdf",sep=""), width = 12, height = 8)
Tal1f6.p1
dev.off()
```


```{r}
plot_loc_tal1 <- "~/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Figures/plots/Fig3/Tal1/"
Tal1.features <- as_tibble(read_clip_tbl(header=FALSE))
```

```{r Plot all Tal1 features}
for (i in 1:nrow(Tal1.features)){
  p <- plotHorizDotplot_v2(feature.coords = as.character(Tal1.features[i,2]))
  if(is.list(p)){
  pdf(file=paste(plot_loc_tal1,paste("Tal1f",i,"_dotplot.pdf",sep=""),sep=""), width = 15, height = 8)
  p3 <- p$p1 + p$p2 + plot_layout(widths = c(5, 1))
  print(p3)
  dev.off()
  }
}

```

```{r}
plot_loc_vsx2 <- "~/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Figures/plots/Fig3/Vsx2/"
Vsx2.features <- as_tibble(read_clip_tbl(header=FALSE))
```

```{r Plot all Vsx2 features}
for (i in 1:nrow(Vsx2.features)){
  p <- plotHorizDotplot_v2(feature.coords = as.character(Vsx2.features[i,2]))
  if(is.list(p)){
  pdf(file=paste(plot_loc_vsx2,paste("Vsx2f",i,"_dotplot.pdf",sep=""),sep=""), width = 15, height = 8)
  p3 <- p$p1 + p$p2 + plot_layout(widths = c(5, 1))
  print(p3)
  dev.off()
  }
}

```
```{r}
plot_loc_gata2 <- "~/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Figures/plots/Fig3/Gata2/"
Gata2.features <- as_tibble(read_clip_tbl(header=FALSE))
```

```{r Plot all Gata2 features}
for (i in 1:nrow(Gata2.features)){
  p <- plotHorizDotplot_v2(feature.coords = as.character(Gata2.features[i,2]))
  if(is.list(p)){
  pdf(file=paste(plot_loc_gata2,paste("Gata2f",i,"_dotplot.pdf",sep=""),sep=""), width = 15, height = 8)
  p3 <- p$p1 + p$p2 + plot_layout(widths = c(5, 1))
  print(p3)
  dev.off()
  }
}

```

```{r}
plot_loc_gata3 <- "~/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Figures/plots/Fig3/Gata3/"
Gata3.features <- as_tibble(read_clip_tbl(header=FALSE))
```

```{r Plot all Gata3 features}
for (i in 1:nrow(Gata3.features)){
  p <- plotHorizDotplot_v2(feature.coords = as.character(Gata3.features[i,2]))
  if(is.list(p)){
  pdf(file=paste(plot_loc_gata3,paste("Gata3f",i,"_dotplot.pdf",sep=""),sep=""), width = 15, height = 8)
    p3 <- p$p1 + p$p2 + plot_layout(widths = c(5, 1))
  print(p3)
  dev.off()
  }
}

```

```{r Tal1f6}
Tal1f6.p1 <- plotHorizDotplot_v2(feature.coords = tal1.feat6)

pdf(file=paste(plot_loc,"Tal1f6_dotplot_v2_abs_nonlog1p.pdf",sep=""), width = 12, height = 8)
Tal1f6.p1
dev.off()
```

```{r Tal1f11}
Tal1f11.p1 <- plotHorizDotplot(feature.coords = tal1.feat11)

pdf(file=paste(plot_loc,"Tal1f11_dotplot.pdf",sep=""), width = 12, height = 8)
Tal1f11.p1
dev.off()
```

```{r Tal1f14}
Tal1f14.p1 <- plotHorizDotplot(feature.coords = tal1.feat14)

pdf(file=paste(plot_loc,"Tal1f14_dotplot.pdf",sep=""), width = 12, height = 8)
Tal1f14.p1
dev.off()
```

```{r Vsx2f5}
Vsx2f5.p1 <- plotHorizDotplot(feature.coords = vsx2.feat5)

pdf(file=paste(plot_loc,"Vsx2f5_dotplot.pdf",sep=""), width = 12, height = 8)
Vsx2f5.p1
dev.off()
```

```{r Vsx2f12}
Vsx2f12.p1 <- plotHorizDotplot(feature.coords = vsx2.feat12)

pdf(file=paste(plot_loc,"Vsx2f12_dotplot.pdf",sep=""), width = 12, height = 8)
Vsx2f12.p1
dev.off()
```

```{r Vsx2f15}
Vsx2f15.p1 <- plotHorizDotplot(feature.coords = vsx2.feat15)

pdf(file=paste(plot_loc,"Vsx2f15_dotplot.pdf",sep=""), width = 12, height = 8)
Vsx2f15.p1
dev.off()
```

```{r Vsx2f17}
Vsx2f17.p1 <- plotHorizDotplot(feature.coords = vsx2.feat17)

pdf(file=paste(plot_loc,"Vsx2f17_dotplot.pdf",sep=""), width = 12, height = 8)
Vsx2f17.p1
dev.off()
```

```{r}
library(igvR)
igv <- igvR()
setBrowserWindowTitle(igv, "mm10 TOBIAS")
setGenome(igv, "mm10")
showGenomicRegion(igv, "TAL1")

TFBS_name <- test.data.filt %>% select(seqnames,start,end,TFBS_name)

TFBS_name <- DataFrameAnnotationTrack("TFBS_name", TFBS_name, color="forestgreen", displayMode="EXPANDED", trackHeight = 500)
displayTrack(igv, TFBS_name)

GA1_2_score <- test.data.filt %>% select(seqnames,start,end,GA1_2_score)

CO1_2_score <- test.data.filt %>% select(seqnames,start,end,CO1_2_score)

GA1_2_score.track <- DataFrameQuantitativeTrack("CO1_2_score", as.data.frame(CO1_2_score), color="red", autoscale=TRUE)
CO1_2_score.track <- DataFrameQuantitativeTrack("GA1_2_score", as.data.frame(GA1_2_score), color="red", autoscale=TRUE)
displayTrack(igv, CO1_2_score.track)
displayTrack(igv, GA1_2_score.track)

GA1_2_exp <- test.data.filt %>% select(seqnames,start,end,GA1_2.x)
GA1_2_exp.track <- DataFrameQuantitativeTrack("GA1_2_exp", as.data.frame(CO1_2_score), color="blue", autoscale=TRUE)
displayTrack(igv, GA1_2_exp.track)

current.loc <- StringToGRanges(paste(getGenomicRegion(igv)$chrom, getGenomicRegion(igv)$start, getGenomicRegion(igv)$end, sep="-"))
cons.filt.subset <- cons.filt[cons.filt %over% current.loc]
cons_filt <- igvR::GRangesQuantitativeTrack("cons_filt",cons.filt.subset)
displayTrack(igv, cons_filt)
```



```{r Reading Seurat dataobject}
dataset <- qread("../scATAC_data/nmm_rV2_subset_relabeled_031023_links.qs", nthreads = 10)
DefaultAssay(dataset) <- "RNA"

dataset$rv2.lineage_re <- case_when(
  dataset$rv2.lineage %in% "PRO1" ~ "PRO1_2",
  dataset$rv2.lineage %in% "PRO2" ~ "PRO1_2",
  dataset$rv2.lineage %in% "GA1" ~ "GA1_2",
  dataset$rv2.lineage %in% "GA2" ~ "GA1_2",
  dataset$rv2.lineage %in% "GA3" ~ "GA3_4",
  dataset$rv2.lineage %in% "GA4" ~ "GA3_4",
  dataset$rv2.lineage %in% "GA5" ~ "GA5_6",
  dataset$rv2.lineage %in% "GA6" ~ "GA5_6",
  dataset$rv2.lineage %in% "CO1" ~ "CO1_2",
  dataset$rv2.lineage %in% "CO2" ~ "CO1_2",
  dataset$rv2.lineage %in% "GL1" ~ "GL1_2",
  dataset$rv2.lineage %in% "GL2" ~ "GL1_2",
  dataset$rv2.lineage %in% "GL3" ~ "GL3_4",
  dataset$rv2.lineage %in% "GL4" ~ "GL3_4",
  dataset$rv2.lineage %in% "GL5" ~ "GL5",
)
```



# H12 Snakemake

```{r Read H12 snakemake TOBIAS results}
#rV2.groups.tobias.h12.gr.dr <- get_BINDetect_snakemake_results_gr("/Volumes/MyBookDuo/Data/TOBIAS_input_output/E12rV2_081223/TFBS/")
# TODO read qs
rV2.groups.tobias.h12.gr.dr <- qread("../analysis/rV2.groups.tobias.h12.gr.dr.regrouped.qs", nthreads = 8)

H12.metadata <- qread(file="../analysis/H12_metadata_mod.qs")

#rV2.h12.overview_re <- read_tsv("/Volumes/MyBookDuo/Data/TOBIAS_input_output/E12rV2_081223/TFBS/bindetect_results.txt", col_names = TRUE)

rV2.h12.overview_re <- read_tsv("../analysis/bindetect_results.txt", col_names = TRUE)

# Adding TF gene names and ensembl ids into the same tibble as TOBIAS overview data from H12 metadata
rV2.h12.overview_re$TF.gene_name <- sapply(pull(rV2.h12.overview_re, motif_id), function(motif){
  getTF.gene_symbol(TF.motif = motif,HOCOMOCO.metadata = H12.metadata, H.version = "H12")
})

rV2.h12.overview_re$ensg_id <- sapply(pull(rV2.h12.overview_re, motif_id), function(motif){
  getTF.id(TF.motif = motif,HOCOMOCO.metadata = H12.metadata, H.version = "H12")
})
```

```{r Plot TOBIAS overview data, fig.height=12}
TF.expression.data.avg <- as.data.frame(AverageExpression(dataset, assays = "RNA", features = pull(rV2.h12.overview_re,ensg_id), group.by = "rv2.lineage_re")[[1]])
colnames(TF.expression.data.avg) <- paste(colnames(TF.expression.data.avg), "_mean_expression",sep="")

rV2.h12.overview_combined<-left_join(rV2.h12.overview_re, rownames_to_column(TF.expression.data.avg,var="ensg_id"))

overview.fp.score.tb <- filter(rV2.h12.overview_combined, !is.na(ensg_id)) %>% select(motif_id, TF.gene_name, ends_with("_mean_score"))
overview.expression.tb <- filter(rV2.h12.overview_combined, !is.na(ensg_id)) %>% select(motif_id, TF.gene_name, ends_with("_mean_expression"))                
# Continue from here
fp.overview.mat <- as.matrix(select(overview.fp.score.tb, ends_with("_score")))
rownames(fp.overview.mat) <- filter(rV2.h12.overview_combined, !is.na(ensg_id)) %>% pull(motif_id)
fp.overview.mat<-t(scale(t(fp.overview.mat)))[,c(1,5,6,7,3,8,2,4)]

expression.overview.mat <- as.matrix(select(overview.expression.tb, ends_with("_expression")))
rownames(expression.overview.mat) <- filter(rV2.h12.overview_combined, !is.na(ensg_id)) %>% pull(motif_id)
expression.overview.mat<-t(scale(t(expression.overview.mat)))[,c(8,1,2,3,4,5,6,7)]

na.ni <- which(apply(expression.overview.mat,1,function(r){all(is.na(r))})==TRUE)
```

```{r Expression average matrix Spearman correlation calculation}
TF.spear <- sapply(1:nrow(fp.overview.mat[-na.ni,]),function(n){
  cor(fp.overview.mat[-na.ni,][n,],expression.overview.mat[-na.ni,][n,], method = "spearman")
})

TF.motifs <- rownames(fp.overview.mat)[-na.ni]
TF.motifs.rep <- paste(TF.motifs,TF.motifs,sep="_")

TF.shifted <- sapply(1:nrow(fp.overview.mat[-na.ni,]),function(n){
  which.max(expression.overview.mat[-na.ni,][n,]) < which.max(fp.overview.mat[-na.ni,][n,])
})

high.cor.ni <- which(TF.spear>.5)
shifted.cor.ni <- which(TF.shifted & TF.spear<0.5)
no.cor.ni <- which(!TF.spear>.5 & !TF.shifted)
```

```{r Table of TF-motif classes}
TF.motifs.out <- tibble(TF.motif=TF.motifs)
TF.motifs.out$expression.class <- case_when(
  TF.motifs %in% TF.motifs[high.cor.ni] ~ "Correlated",
  TF.motifs %in% TF.motifs[shifted.cor.ni] ~ "Shifted",
  TF.motifs %in% TF.motifs[no.cor.ni] ~ "Uncorrelated"
)

write_csv(TF.motifs.out, file = "~/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Tables/TF_expression_classes.csv")
pdf(file=paste(plot_loc,"TF_categories.pdf",sep=""), width = 10, height = 10)
ggpubr::ggbarplot(data=as.data.frame(table(TF.motifs.out$expression.class)),x="Var1",y="Freq", fill="Var1") +xlab("Category")+ylab("Count")
dev.off()
```

```{r Combined plots correlating, fig.height=16, fig.width=10}
combined.matrix <- cbind(expression.overview.mat[-na.ni,][high.cor.ni,],fp.overview.mat[-na.ni,][high.cor.ni,])
combined.matrix.fviz_nbclust <- fviz_nbclust(combined.matrix, FUN=kmeans, method="silhouette")
opt.kmeans <-  which.max(combined.matrix.fviz_nbclust$data$y)

row_ha = rowAnnotation(corr = TF.spear[high.cor.ni])

h.cor <- Heatmap(combined.matrix, cluster_columns = FALSE, cluster_rows = TRUE, show_row_names = TRUE, row_names_gp=gpar(fontsize=5),right_annotation = row_ha, row_split = opt.kmeans, column_split = factor(c(rep("expression",8),rep("fp",8)), levels=c("expression","fp")), column_gap = unit(5, "mm"))

h.cor
pdf(file=paste(plot_loc,"TF_cor_heatmap.pdf",sep=""), width = 10, height = 22)
h.cor
dev.off()
```

```{r Combined plots shifted, fig.height=20, fig.width=10}
combined.matrix <- cbind(expression.overview.mat[-na.ni,][shifted.cor.ni,],fp.overview.mat[-na.ni,][shifted.cor.ni,])
combined.matrix.fviz_nbclust <- fviz_nbclust(combined.matrix, FUN=kmeans, method="silhouette")
opt.kmeans <-  which.max(combined.matrix.fviz_nbclust$data$y)

row_ha = rowAnnotation(corr = TF.spear[shifted.cor.ni])

h.shifted <- Heatmap(combined.matrix, cluster_columns = FALSE, cluster_rows = TRUE, show_row_names = TRUE, row_names_gp=gpar(fontsize=4),right_annotation = row_ha, row_split = opt.kmeans, column_split = factor(c(rep("expression",8),rep("fp",8)), levels=c("expression","fp")), column_gap = unit(5, "mm"))

h.shifted
pdf(file=paste(plot_loc,"TF_shifted_heatmap.pdf",sep=""), width = 10, height = 22)
h.shifted
dev.off()
```

```{r formFootprintMatrix Tal1_feat6}
foot.score.GA.tal1_6 <- formFootprintMatrix.overConditions(rV2.groups.tobias.h12.gr.dr, conditions = c("PRO1_2","CO1_2","GA1_2","GA3_4","GA5_6"), gr.filter = tal1.feat6, na.omit=TRUE, binary=FALSE)
colnames(foot.score.GA.tal1_6$footprint.matrix) <- str_remove(string=colnames(foot.score.GA.tal1_6$footprint.matrix), pattern = "_.*")

foot.score.GL.tal1_6 <- formFootprintMatrix.overConditions(rV2.groups.tobias.h12.gr.dr, conditions = c("PRO1_2","CO1_2","GL1_2","GL3_4","GL5"), gr.filter = tal1.feat6, na.omit=TRUE, binary=FALSE)
colnames(foot.score.GL.tal1_6$footprint.matrix) <- str_remove(string=colnames(foot.score.GL.tal1_6$footprint.matrix), pattern = "_.*")
```

```{r Plot footprint matrix Tal1_feat6 GABA branch correlating and shifted, fig.width=10, fig.height=16}
#common.motifs <- intersect(TF.motifs[high.cor.ni],colnames(foot.score.GA.tal1_6$footprint.matrix))
common.motifs <- unique(c(intersect(TF.motifs[high.cor.ni],colnames(foot.score.GA.tal1_6$footprint.matrix)),intersect(TF.motifs[shifted.cor.ni],colnames(foot.score.GA.tal1_6$footprint.matrix))))

foot.score.GA.tal1_6.cor <- foot.score.GA.tal1_6
foot.score.GA.tal1_6.cor$footprint.matrix <- foot.score.GA.tal1_6.cor$footprint.matrix[,common.motifs]

p1 <- plotTFfootprint.heatmap(foot.score.GA.tal1_6.cor, filter.unbound = TRUE, with.expression = TRUE, Seurat.dataset = dataset, HOCOMOCO.metadata = H12.metadata,filter.no.expressed.TF=FALSE, group.by = "rv2.lineage_re")

pdf(file=paste(plot_loc,"Tal1f6_GA_heatmap.pdf",sep=""), width = 10, height = 16)
p1
dev.off()
```

```{r Plot footprint matrix Tal1_feat6 GL branch correlating and shifted, fig.width=10, fig.height=16}
common.motifs <- unique(c(intersect(TF.motifs[high.cor.ni],colnames(foot.score.GL.tal1_6$footprint.matrix)),intersect(TF.motifs[shifted.cor.ni],colnames(foot.score.GL.tal1_6$footprint.matrix))))

foot.score.GL.tal1_6.cor <- foot.score.GL.tal1_6
foot.score.GL.tal1_6.cor$footprint.matrix <- foot.score.GL.tal1_6.cor$footprint.matrix[,common.motifs]

p2 <- plotTFfootprint.heatmap(foot.score.GL.tal1_6.cor, filter.unbound = TRUE, with.expression = TRUE, Seurat.dataset = dataset, HOCOMOCO.metadata = H12.metadata,filter.no.expressed.TF=FALSE, group.by = "rv2.lineage_re")
pdf(file=paste(plot_loc,"Tal1f6_GL_heatmap.pdf",sep=""), width = 10, height = 16)
p2
dev.off()
```

```{r log2fc of conditions GA1_2 and GL1_2 vs CO1_"}
# tmp.CO <- unlist(sapply(TF.motifs.rep,function(tf){get.fp.logfc(tobias.gr = rV2.groups.tobias.h12.gr.dr, TF.motif = tf, cond1="PRO1_2",cond2="CO1_2", gr.filter = tal1.feat6)}))
# CO.tb <- tibble(group="CO",motif=names(tmp.CO),fp.log2fc=tmp.CO)
# CO.tb$motif <- str_remove(CO.tb$motif, pattern = "_.*")
# 
# CO.tb$TF.exp.log2fc <- sapply(CO.tb$motif,function(m){
#   get.gene.exp.log2fc(dataset=dataset, motif=m,cond1="PRO1_2",cond2="CO1_2",group.by = "rv2.lineage_re", H.metadata = H12.metadata)
# })
# 
# tmp.GA <- unlist(sapply(TF.motifs.rep,function(tf){get.fp.logfc(tobias.gr = rV2.groups.tobias.h12.gr.dr, TF.motif = tf, cond1="CO1_2",cond2="GA1_2", gr.filter = tal1.feat6)}))
# GA.tb <- tibble(group="GA",motif=names(tmp.GA),fp.log2fc=tmp.GA)
# GA.tb$motif <- str_remove(GA.tb$motif, pattern = "_.*")
# 
# GA.tb$TF.exp.log2fc <- sapply(GA.tb$motif,function(m){
#   get.gene.exp.log2fc(dataset=dataset, motif=m,cond1="CO1_2",cond2="GA1_2",group.by = "rv2.lineage_re", H.metadata = H12.metadata)
# })
# 
# 
# tmp.GL <- unlist(sapply(TF.motifs.rep,function(tf){get.fp.logfc(tobias.gr = rV2.groups.tobias.h12.gr.dr, TF.motif = tf, cond1="CO1_2",cond2="GL1_2", gr.filter = tal1.feat6)}))
# GL.tb <- tibble(group="GL",motif=names(tmp.GL),fp.log2fc=tmp.GL)
# GL.tb$motif <- str_remove(GL.tb$motif, pattern = "_.*")
# 
# GL.tb$TF.exp.log2fc <- sapply(GL.tb$motif,function(m){
#   get.gene.exp.log2fc(dataset=dataset, motif=m,cond1="CO1_2",cond2="GL1_2",group.by = "rv2.lineage_re", H.metadata = H12.metadata)
# })
# 
# 
# GA_GL.tal1f6 <- full_join(GA.tb,GL.tb)
# GA_GL.tal1f6 <- full_join(GA_GL.tal1f6,CO.tb)
# #GA_GL.tal1f6$motif <- str_remove(GA_GL.tal1f6$motif, pattern = "_.*")
# GA_GL.tal1f6$fp.log2fc <- GA_GL.tal1f6$fp.log2fc*-1
# 
# # GA_GL.tal1f6$TF.exp.log2fc <- sapply(GA_GL.tal1f6$motif,function(m){
# #   get.gene.exp.log2fc(dataset=dataset, motif=m,cond1="CO1_2",cond2="GA1_2",group.by = "rv2.lineage_re", H.metadata = H12.metadata)
# # })
# 
# GA_GL.tal1f6<- left_join(GA_GL.tal1f6, TF.motifs.out, by=c("motif"="TF.motif"))
# 
# GA_GL.tal1f6$TF.exp.log2fc <- GA_GL.tal1f6$TF.exp.log2fc*-1
```


```{r Plot CO1_2 to GA1_2 scatter plot, fig.height=10, fig.width=10}
# ggplot(filter(GA_GL.tal1f6, group=="GA" & expression.class %in% c("Shifted","Correlated")), aes(y=fp.log2fc, x=TF.exp.log2fc)) + geom_point() + theme_minimal() + ggrepel::geom_label_repel(aes(label = ifelse((fp.log2fc > 1.5 & TF.exp.log2fc > 1.5), motif, "")),size = 2.5)
```

```{r Dotplot version of the above, fig.height=10, fig.width=6}
# ggplot() + 
# geom_point(aes(x = group, y=motif, size=fp.log2fc, colour=TF.exp.log2fc, shape=expression.class),filter(GA_GL.tal1f6, fp.log2fc > .5 & group=="CO" & TF.exp.log2fc > 0.5 & !(expression.class=="Uncorrelated"))) +
# theme_minimal() +
# geom_point(aes(x = group, y=motif, size=fp.log2fc, colour=TF.exp.log2fc, shape=expression.class),filter(GA_GL.tal1f6, fp.log2fc > .5 & group=="GA" & TF.exp.log2fc > 0.5 & !(expression.class=="Uncorrelated"))) +
# theme_minimal() + scale_colour_gradient2(low = "grey", mid="yellow",high = "red") +
# geom_point(aes(x = group, y=motif, size=fp.log2fc, colour=TF.exp.log2fc, shape=expression.class),filter(GA_GL.tal1f6, fp.log2fc > .5  & group=="GL" & TF.exp.log2fc >.5 & !(expression.class=="Uncorrelated"))) +
# theme_minimal()
```

```{r comp dotplots of Tal1}
Tal1.f6 <- plotFootprintDotplot.rV2.comp(tobias.fp.gr = rV2.groups.tobias.h12.gr.dr, gr.filter = tal1.feat6, TF.motifs = TF.motifs.out, dataset=dataset)

pdf(file=paste(plot_loc,"Tal1_f6_comp_dotplot.pdf",sep=""), width = 8, height = 16)
Tal1.f6
dev.off()

Tal1.f11 <- plotFootprintDotplot.rV2.comp(tobias.fp.gr = rV2.groups.tobias.h12.gr.dr, gr.filter = tal1.feat11, TF.motifs = TF.motifs.out, dataset=dataset)

pdf(file=paste(plot_loc,"Tal1_f11_comp_dotplot.pdf",sep=""), width = 8, height = 16)
Tal1.f11
dev.off()

Tal1.f14 <- plotFootprintDotplot.rV2.comp(tobias.fp.gr = rV2.groups.tobias.h12.gr.dr, gr.filter = tal1.feat14, TF.motifs = TF.motifs.out, dataset=dataset)

pdf(file=paste(plot_loc,"Tal1_f14_comp_dotplot.pdf",sep=""), width = 8, height = 16)
Tal1.f14
dev.off()
```


```{r comp dotplots of Vsx2}
Vsx2.feat5.p1 <- plotFootprintDotplot.rV2(tobias.fp.gr = rV2.groups.tobias.h12.gr.dr, gr.filter = vsx2.feat5, TF.motifs = TF.motifs.out, dataset=dataset, parallel=TRUE)

pdf(file=paste(plot_loc,"Vsx2_f5_comp_dotplot.pdf",sep=""), width = 8, height = 16)
Vsx2.feat5.p1
dev.off()

Vsx2.feat12.p1 <- plotFootprintDotplot.rV2(tobias.fp.gr = rV2.groups.tobias.h12.gr.dr, gr.filter = vsx2.feat12, TF.motifs = TF.motifs.out, dataset=dataset, parallel=T)

pdf(file=paste(plot_loc,"Vsx2_feat12_comp_dotplot.pdf",sep=""), width = 8, height = 16)
Vsx2.feat12.p1
dev.off()

Vsx2.feat15.p1 <- plotFootprintDotplot.rV2(tobias.fp.gr = rV2.groups.tobias.h12.gr.dr, gr.filter = vsx2.feat15, TF.motifs = TF.motifs.out, dataset=dataset, parallel = T)

pdf(file=paste(plot_loc,"Vsx2_feat15_comp_dotplot.pdf",sep=""), width = 8, height = 16)
Vsx2.feat15.p1
dev.off()

Vsx2.feat17.p1 <- plotFootprintDotplot.rV2(tobias.fp.gr = rV2.groups.tobias.h12.gr.dr, gr.filter = vsx2.feat17, TF.motifs = TF.motifs.out, dataset=dataset, parallel = T)

pdf(file=paste(plot_loc,"Vsx2_feat17_comp_dotplot.pdf",sep=""), width = 8, height = 16)
Vsx2.feat17.p1
dev.off()
```

```{r non_comp dotplots of Tal1}
Tal1.f6 <- plotFootprintDotplot.rV2.noncomp(tobias.fp.gr = rV2.groups.tobias.h12.gr.dr, gr.filter = tal1.feat6, TF.motifs = TF.motifs.out, dataset=dataset, mc.cores=9, cons.gr=cons.gr)

pdf(file=paste(plot_loc,"Tal1_f6_non_comp_dotplot.pdf",sep=""), width = 8, height = 20)
Tal1.f6$plot
dev.off()

Tal1.f11 <- plotFootprintDotplot.rV2.noncomp(tobias.fp.gr = rV2.groups.tobias.h12.gr.dr, gr.filter = tal1.feat11, TF.motifs = TF.motifs.out, dataset=dataset, mc.cores=9, cons.gr=cons.gr)

pdf(file=paste(plot_loc,"Tal1_f11_non_comp_dotplot.pdf",sep=""), width = 8, height = 20)
Tal1.f11$plot
dev.off()

Tal1.f14 <- plotFootprintDotplot.rV2.noncomp(tobias.fp.gr = rV2.groups.tobias.h12.gr.dr, gr.filter = tal1.feat14, TF.motifs = TF.motifs.out, dataset=dataset, mc.cores=9, cons.gr=cons.gr)

pdf(file=paste(plot_loc,"Tal1_f14_non_comp_dotplot.pdf",sep=""), width = 8, height = 20)
Tal1.f14$plot
dev.off()
```


```{r non_comp dotplots of Vsx2}
Vsx2.feat5.p1 <- plotFootprintDotplot.rV2.noncomp(tobias.fp.gr = rV2.groups.tobias.h12.gr.dr, gr.filter = vsx2.feat5, TF.motifs = TF.motifs.out, dataset=dataset, cons.gr=cons.gr, mc.cores = 9)

pdf(file=paste(plot_loc,"Vsx2_f5_non_comp_dotplot.pdf",sep=""), width = 8, height = 20)
Vsx2.feat5.p1
dev.off()

Vsx2.feat12.p1 <- plotFootprintDotplot.rV2.noncomp(tobias.fp.gr = rV2.groups.tobias.h12.gr.dr, gr.filter = vsx2.feat12, TF.motifs = TF.motifs.out, dataset=dataset,  cons.gr=cons.gr, mc.cores = 9)

pdf(file=paste(plot_loc,"Vsx2_feat12_non_comp_dotplot.pdf",sep=""), width = 8, height = 20)
Vsx2.feat12.p1
dev.off()

Vsx2.feat15.p1 <- plotFootprintDotplot.rV2.noncomp(tobias.fp.gr = rV2.groups.tobias.h12.gr.dr, gr.filter = vsx2.feat15, TF.motifs = TF.motifs.out, dataset=dataset,  cons.gr=cons.gr, mc.cores = 9)

pdf(file=paste(plot_loc,"Vsx2_feat15_non_comp_dotplot.pdf",sep=""), width = 8, height = 20)
Vsx2.feat15.p1
dev.off()

Vsx2.feat17.p1 <- plotFootprintDotplot.rV2.noncomp(tobias.fp.gr = rV2.groups.tobias.h12.gr.dr, gr.filter = vsx2.feat17, TF.motifs = TF.motifs.out, dataset=dataset,  cons.gr=cons.gr, mc.cores = 9)

pdf(file=paste(plot_loc,"Vsx2_feat17_non_comp_dotplot.pdf",sep=""), width = 8, height = 20)
Vsx2.feat17.p1
dev.off()
```
