---
title: "R Notebook"
output: html_notebook
---

```{r}
library(netZooR)
library(qs)
library(Seurat)
library(Signac)
library(GenomicRanges)
library(tidyverse)
library(hash)
library(DBI)
library(gplots)
cores<-8
```

```{r}
con <- DBI::dbConnect(RSQLite::SQLite(), dbname = "~/Workspace/TOBIAS.dr.h12_230424.sqlite")
options(timeout = 30000)
```


```{r Reading Seurat dataobject}
dataset <- qread("../scATAC_data/nmm_rV2_subset_relabeled_031023_links.qs", nthreads = cores)
DefaultAssay(dataset) <- "RNA"

dataset$rv2.lineage_re <- case_when(
  dataset$rv2.lineage %in% "PRO1" ~ "PRO1_2",
  dataset$rv2.lineage %in% "PRO2" ~ "PRO1_2",
  dataset$rv2.lineage %in% "GA1" ~ "GA1_2",
  dataset$rv2.lineage %in% "GA2" ~ "GA1_2",
  dataset$rv2.lineage %in% "GA3" ~ "GA3_4",
  dataset$rv2.lineage %in% "GA4" ~ "GA3_4",
  dataset$rv2.lineage %in% "GA5" ~ "GA5_6",
  dataset$rv2.lineage %in% "GA6" ~ "GA5_6",
  dataset$rv2.lineage %in% "CO1" ~ "CO1_2",
  dataset$rv2.lineage %in% "CO2" ~ "CO1_2",
  dataset$rv2.lineage %in% "GL1" ~ "GL1_2",
  dataset$rv2.lineage %in% "GL2" ~ "GL1_2",
  dataset$rv2.lineage %in% "GL3" ~ "GL3_4",
  dataset$rv2.lineage %in% "GL4" ~ "GL3_4",
  dataset$rv2.lineage %in% "GL5" ~ "GL5",
)

DefaultAssay(dataset) <- "peaks"
all.features <- StringToGRanges(rownames(dataset))

gene_id2name <- hash(rownames(dataset[['RNA']][[]]),dataset[['RNA']][[]][,1])
gene_name2id <- hash(dataset[['RNA']][[]][,1],rownames(dataset[['RNA']][[]]))
```

```{r Fetch data layers for graph building}
DefaultAssay(dataset) <- "RNA"
exp.data <- FetchData(object = dataset, vars = rownames(dataset))
colnames(exp.data) <- str_to_upper(sapply(colnames(exp.data),function(id){gene_id2name[[id]]}))
```

```{r Define cell groups and linked features}
lineage.li <- dataset$rv2.lineage_re %in% c("CO1_2","GA1_2")

design <- dataset$rv2.lineage_re[dataset$rv2.lineage_re %in% c("CO1_2","GA1_2")]
design.bin <- ifelse(design=="CO1_2",0,1)

# First run
# linked.features <- as_tibble(dbGetQuery(con, 'SELECT feature FROM links as li WHERE ABS(li.zscore)>2 AND li.pvalue<0.001;')) %>% pull(feature)
# linked.genes <- unique(as_tibble(dbGetQuery(con, 'SELECT ensg_id FROM links as li WHERE ABS(li.zscore)>2 AND li.pvalue<0.001;')) %>% pull(ensg_id))
# TF.2.gene.data <- as_tibble(dbGetQuery(con, 'SELECT tb.TF_gene_name, gm.gene_name, li.zscore FROM links as li, tobias as tb, gene_metadata as gm WHERE ABS(li.zscore)>2 AND li.pvalue<0.001 AND tb.features=li.feature AND gm.ensg_id=li.ensg_id;'))

# Second run
linked.features <- as_tibble(dbGetQuery(con, 'SELECT feature FROM links as li WHERE li.zscore>2 AND li.pvalue<0.001;')) %>% pull(feature)
linked.genes <- unique(as_tibble(dbGetQuery(con, 'SELECT ensg_id FROM links as li WHERE li.zscore>2 AND li.pvalue<0.001;')) %>% pull(ensg_id))
TF.2.gene.data <- as_tibble(dbGetQuery(con, 'SELECT tb.TF_gene_name, gm.gene_name, li.zscore FROM links as li, tobias as tb, gene_metadata as gm WHERE li.zscore>2 AND li.pvalue<0.001 AND tb.features=li.feature AND gm.ensg_id=li.ensg_id AND (tb.GA1_2_bound=1 OR tb.CO1_2_bound=1);'))

TF.2.gene.data$gene_name <- str_to_upper(TF.2.gene.data$gene_name)
```

```{r Subset datamatrices}
exp.lineage <- exp.data[lineage.li,intersect(unique(c(str_to_upper(TF.2.gene.data$gene_name),TF.2.gene.data$TF_gene_name)),colnames(exp.data))]
zero.filt <- which(colSums(exp.lineage>0)==0)
exp.lineage <- t(exp.lineage[,-zero.filt])


TF.gene.data.sum <- TF.2.gene.data %>% filter(TF_gene_name %in% str_to_upper(rownames(exp.lineage)) | gene_name %in% str_to_upper(rownames(exp.lineage))) %>% group_by(TF_gene_name, gene_name) %>% summarize(avg.score=mean(zscore)) %>% filter(avg.score>0) %>% ungroup()

TF.gene.data.sum$avg.score <- TF.gene.data.sum$avg.score/max(TF.gene.data.sum$avg.score)
```

```{r Run Monster, warning=FALSE}
#yeast$exp.cc[is.na(yeast$exp.cc)] <- mean(as.matrix(yeast$exp.cc),na.rm=TRUE)
monsterRes <- monster(expr=exp.lineage, design=design.bin, motif=as.data.frame(TF.gene.data.sum), nullPerms=100, numMaxCores=12, alphaw=1, outputDir = "../analysis/Monster_2/")
saveRDS(monsterRes, file="../analysis/Monster_2.Rds")
```

```{r fig.width=20, fig.height=8}
monsterPlotMonsterAnalysis(monsterRes, rescale='significance') + theme(axis.text.x = element_text(size=7))
```

```{r}
heatmap.2(slot(monsterRes, 'tm'), col = "bluered",
density.info="none",  
trace="none", 
dendrogram='none',     
Rowv=FALSE,
Colv=FALSE)
```


```{r, fig.width=20, fig.height=20}
monsterTransitionNetworkPlot(monsterRes, numEdges=150,numTopTFs = 150)
```
```{r, fig.height=18, fig.width=18}
monsterTransitionPCAPlot(monsterRes)
```



```{r}
sessionInfo()
```

