---
title: "R Notebook for selector gene target analysis in rV2"
author: Sami Kilpinen
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output: 
  html_notebook:
    code_folding: hide
---

```{r Local settings}
source("local_settings.R")
```

```{r Packages, include=FALSE}
library(tidyverse)
library(Seurat)
library(Signac)
library(qs)
library(readxl)
library(hash)
library(fgsea)
library(DBI)
library(hash)
library(biomaRt)
library(presto)
library(ggpubr)
library(purrr)
library(patchwork)
library(parallel)
library(umap)
library(topGO)
library(plotly)
library(dplyr)
library(cowplot)
library(ggrepel)
library(ComplexHeatmap)
library(future)
library(circlize)
library(GenomicRanges)
source(paste(auxcode.path,"AuxFunctions.R",sep=""))
source(paste(auxcode.path,"TargetAnalysis.R",sep=""))
```

```{r}
plan("multisession",workers=cores)
```

```{r Setting DBI options, include=FALSE}
con <- DBI::dbConnect(RSQLite::SQLite(), dbname=paste(db.path,dbname.rV2,sep=""))
options(timeout = 30000)
```

```{r}
plot_loc<-"~/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Figures/plots/Fig6/"
```

# Target counts {.tabset}

## Counts with C&T True and Cons True FT True in rV2

```{r, eval=FALSE}
TF_names <- c("Tal1","Gata2","Gata3","Vsx2")
targetGenes.set <- lapply(TF_names,getTargets_rV2,con=con)
names(targetGenes.set) <- TF_names
sapply(targetGenes.set,function(p){length(p$targets_gene_name)})
```

## Feature level targets

```{r}
TF_names <- c("Tal1","Gata2","Gata3","Vsx2","Insm1","Ebf1","Tead2")
targetGenes.set <- lapply(TF_names,getTargets_rV2_feature_level,con=con)
names(targetGenes.set) <- TF_names
sapply(targetGenes.set,function(p){length(p$targets_gene_name)})
```

```{r Adding C&T from Gata2/3 for Tal1 targets}
Tal1_target_granges <- StringToGRanges(targetGenes.set$Tal1$target.data$feature, sep = c("-", "-"))
query <- 'SELECT chr, start, end FROM ct_data WHERE target_gene_name = "Gata2"'
Gata2_cnt <- as_tibble(dbGetQuery(con, query),.name_repair = "unique")
Gata2_granges <- makeGRangesFromDataFrame(Gata2_cnt)
query <- 'SELECT chr, start, end FROM ct_data WHERE target_gene_name = "Gata3"'
Gata3_cnt <- as_tibble(dbGetQuery(con, query),.name_repair = "unique")
Gata3_granges <- makeGRangesFromDataFrame(Gata3_cnt)
gata2_overlap <- countOverlaps(Tal1_target_granges, Gata2_granges, minoverlap = 1)
targetGenes.set[["Tal1"]][["target.data"]][["Gata2_ct_overlap"]] <- replace(gata2_overlap, which(gata2_overlap > 0), 1)
gata3_overlap <- countOverlaps(Tal1_target_granges, Gata3_granges, minoverlap = 1)
targetGenes.set[["Tal1"]][["target.data"]][["Gata3_ct_overlap"]] <- replace(gata3_overlap, which(gata3_overlap > 0), 1)
```

# Expression analysis of target genes {.tabset}

```{r Reading rV2 data, include=FALSE, eval=TRUE}
rV2.dataset <- qread("../scATAC_data/nmm_rV2_subset_relabeled_031023_links.qs", nthreads = cores)

rV2.dataset$rv2.lineage_re <- case_when(
  rV2.dataset$rv2.lineage %in% "PRO1" ~ "PRO1_2",
  rV2.dataset$rv2.lineage %in% "PRO2" ~ "PRO1_2",
  rV2.dataset$rv2.lineage %in% "GA1" ~ "GA1_2",
  rV2.dataset$rv2.lineage %in% "GA2" ~ "GA1_2",
  rV2.dataset$rv2.lineage %in% "GA3" ~ "GA3_4",
  rV2.dataset$rv2.lineage %in% "GA4" ~ "GA3_4",
  rV2.dataset$rv2.lineage %in% "GA5" ~ "GA5_6",
  rV2.dataset$rv2.lineage %in% "GA6" ~ "GA5_6",
  rV2.dataset$rv2.lineage %in% "CO1" ~ "CO1_2",
  rV2.dataset$rv2.lineage %in% "CO2" ~ "CO1_2",
  rV2.dataset$rv2.lineage %in% "GL1" ~ "GL1_2",
  rV2.dataset$rv2.lineage %in% "GL2" ~ "GL1_2",
  rV2.dataset$rv2.lineage %in% "GL3" ~ "GL3_4",
  rV2.dataset$rv2.lineage %in% "GL4" ~ "GL3_4",
  rV2.dataset$rv2.lineage %in% "GL5" ~ "GL5",
)

DefaultAssay(rV2.dataset) <- "RNA"

gene_id2name <- hash(rownames(rV2.dataset[['RNA']][[]]),rV2.dataset[['RNA']][[]][,1])
gene_name2id <- hash(rV2.dataset[['RNA']][[]][,1],rownames(rV2.dataset[['RNA']][[]]))
qsave(gene_id2name,file="gene_id2name.qs")
qsave(gene_name2id,file="gene_name2id.qs")
```

```{r Calculating averages and DEG for GA/GL axis, include=FALSE}
assay.averages.rV2 <- AverageExpression(rV2.dataset,group.by = "rv2.lineage_re")

Idents(rV2.dataset) <- rV2.dataset$rv2.lineage_re
DEG.res.1 <- FindMarkers(rV2.dataset, ident.1="GA1_2", ident.2 = "GL1_2", logfc.threshold=0, assay = "RNA", slot="data", mean.fxn=mean.fxn) %>% rownames_to_column(var="ensg_id") %>% as_tibble()

colnames(assay.averages.rV2$RNA) <- str_replace_all(colnames(assay.averages.rV2$RNA), pattern = "-", replacement = "_")

DEG.res.1$GL1_2_exp_avg_log2 <- log(assay.averages.rV2$RNA[DEG.res.1$ensg_id,"GL1_2"]+1,base=2)
DEG.res.1$GA1_2_exp_avg_log2 <- log(assay.averages.rV2$RNA[DEG.res.1$ensg_id,"GA1_2"]+1, base=2)
DEG.res.1$GL1_2_exp_avg <- assay.averages.rV2$RNA[DEG.res.1$ensg_id,"GL1_2"]
DEG.res.1$GA1_2_exp_avg <- assay.averages.rV2$RNA[DEG.res.1$ensg_id,"GA1_2"]
DEG.res.1$gene_name <- sapply(DEG.res.1$ensg_id,function(ensg_id){hash::values(gene_id2name[ensg_id])})
DEG.res.1 <- DEG.res.1 %>% mutate(across(where(is.numeric), \(x) round(x, digits = 8))) %>% arrange(avg_log2FC)
```

```{r Correlations between target genes and selector genes in rV2, warning=FALSE, eval=FALSE}
TF_names <- c("Tal1","Gata2","Gata3","Vsx2","Insm1","Ebf1","Tead2")

PRO_GA.subset <- subset(rV2.dataset, idents=c("PRO1_2","GA1_2"))
PRO_GL.subset <- subset(rV2.dataset, idents=c("PRO1_2","GL1_2"))

# CO_GA.correlations 
expression_matrix <- GetAssayData(PRO_GA.subset, slot = "data")
TF.cors.PRO_GA <- lapply(TF_names,function(TF){
    correlations <- mclapply(1:nrow(expression_matrix), function(i) {
      gene_expr <- expression_matrix[i,]
      if (sd(gene_expr) == 0) {
        return(NA)
      } else {
        cor(expression_matrix[gene_name2id[[TF]], ], gene_expr, method = "spearman")
      }
    }, mc.cores=8)
    names(correlations) <- rownames(expression_matrix)
    return(correlations)
})
names(TF.cors.PRO_GA) <- TF_names

# CO_GL.correlations 
expression_matrix <- GetAssayData(PRO_GL.subset, slot = "data")
TF.cors.PRO_GL <- lapply(TF_names,function(TF){
    correlations <- mclapply(1:nrow(expression_matrix), function(i) {
      gene_expr <- expression_matrix[i,]
      if (sd(gene_expr) == 0) {
        return(NA)
      } else {
        cor(expression_matrix[gene_name2id[[TF]], ], gene_expr, method = "spearman")
      }
    }, mc.cores=8)
    names(correlations) <- rownames(expression_matrix)
    return(correlations)
})
names(TF.cors.PRO_GL) <- TF_names

qsave(TF.cors.PRO_GA,paste(data.path,"TF.cors.PRO_GA.qs",sep=""), nthreads = cores)
qsave(TF.cors.PRO_GL,paste(data.path,"TF.cors.PRO_GL.qs",sep=""), nthreads = cores)
```

```{r Loading precalculated correlations rV2}
TF.cors.PRO_GA <- qread(paste(data.path,"TF.cors.PRO_GA.qs",sep=""), nthreads = cores)
TF.cors.PRO_GL <- qread(paste(data.path,"TF.cors.PRO_GL.qs",sep=""), nthreads = cores)
```

## rV2 target tables with GA1_2 and GL1_2 information

```{r Adding DEG expression correlation to target data tables}
DEG.res.subset <- lapply(targetGenes.set, function(p){
  full_join(filter(DEG.res.1, ensg_id %in% p$targets_gene.id) %>% dplyr::select(-gene_name),p$target.data, by=c("ensg_id"="ensg_id"))
  })

combined_tibble <- map_dfr(DEG.res.subset, ~ .x, .id = "TF")

combined_tibble <- combined_tibble %>%
  mutate(
    GA1_2_exp_avg = assay.averages.rV2$RNA[match(ensg_id, rownames(assay.averages.rV2$RNA)), "GA1_2"],
    GL1_2_exp_avg = assay.averages.rV2$RNA[match(ensg_id, rownames(assay.averages.rV2$RNA)), "GL1_2"],
    GA1_2_exp_avg_log2 = log(assay.averages.rV2$RNA[match(ensg_id, rownames(assay.averages.rV2$RNA)), "GA1_2"]+1, base=2),
    GL1_2_exp_avg_log2 = log(assay.averages.rV2$RNA[match(ensg_id, rownames(assay.averages.rV2$RNA)), "GL1_2"]+1,base = 2)
  )

combined_tibble <- combined_tibble %>%
  rowwise() %>%
  mutate(TF.cor.PRO.GA = TF.cors.PRO_GA[[TF]][[ensg_id]]) %>%
  ungroup()

combined_tibble <- combined_tibble %>%
  rowwise() %>%
  mutate(TF.cor.PRO.GL = TF.cors.PRO_GL[[TF]][[ensg_id]]) %>%
  ungroup()

combined_tibble <- combined_tibble %>% mutate_if(is.numeric, round, digits=3)

combined_tibble <- combined_tibble %>%
  mutate(linkage_kbp = linkage_dist / 1000)

#Tal1.targets <- add_TE_to_tibble(filter(combined_tibble, TF=="Tal1"),te.results[["Tal1"]])

create_dt(filter(combined_tibble, TF=="Tal1"))
create_dt(filter(combined_tibble, TF=="Gata2"))
create_dt(filter(combined_tibble, TF=="Gata3"))
create_dt(filter(combined_tibble, TF=="Vsx2"))
create_dt(filter(combined_tibble, TF=="Insm1"))
create_dt(filter(combined_tibble, TF=="Ebf1"))
create_dt(filter(combined_tibble, TF=="Tead2"))
```

# Target gene expression

```{r Expression and averages rV2}
assay.averages.rV2 <- AverageExpression(rV2.dataset,group.by = "rv2.lineage")

tmp.subset <- subset(rV2.dataset, idents=c("CO1_2","GA1_2"))
assay.rna.GABA <- FetchData(tmp.subset, vars = rownames(tmp.subset), cells = colnames(tmp.subset))

tmp.subset <- subset(rV2.dataset, idents=c("CO1_2","GL1_2"))
assay.rna.GLUT <- FetchData(tmp.subset, vars = rownames(tmp.subset), cells = colnames(tmp.subset))
```


```{r}
plot_heatmap_with_variability(
    gene_list = Tal1.causal.targets, 
    combined_tibble = combined_tibble, 
    rna_data = assay.averages.rV2$RNA, 
    gene_id2name = gene_id2name,
    TF = "Tal1", 
    zscore_condition = "> 0",  # Specify zscore condition
    custom_filter = "& Gata2_ct_overlap == 0 & Gata3_ct_overlap == 0",  # Example of custom filter
    output_file = "~/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Figures/plots/Tal1.targets.pos.Tal.spec.pdf"
)


plot_heatmap_with_variability(
    gene_list = Tal1.causal.targets, 
    combined_tibble = combined_tibble, 
    rna_data = assay.averages.rV2$RNA, 
    gene_id2name = gene_id2name,
    TF = "Tal1", 
    zscore_condition = "> 0",  # Specify zscore condition
    custom_filter = "& (Gata2_ct_overlap==1 | Gata3_ct_overlap==1)",  # Example of custom filter
    output_file = "~/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Figures/plots/Tal1.targets.pos.Tal1_Gata.pdf"
)


plot_heatmap_with_variability(
    gene_list = Tal1.causal.targets, 
    combined_tibble = combined_tibble, 
    rna_data = assay.averages.rV2$RNA, 
    gene_id2name = gene_id2name,
    TF = "Tal1", 
    zscore_condition = "< 0",  # Specify zscore condition
    custom_filter = "& Gata2_ct_overlap==0 & Gata3_ct_overlap==0",  # Example of custom filter
    output_file = "~/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Figures/plots/Tal1.targets.neg.Tal1.spec.pdf"
)


plot_heatmap_with_variability(
    gene_list = Tal1.causal.targets, 
    combined_tibble = combined_tibble, 
    rna_data = assay.averages.rV2$RNA, 
    gene_id2name = gene_id2name,
    TF = "Tal1", 
    zscore_condition = "< 0",  # Specify zscore condition
    custom_filter = " & (Gata2_ct_overlap==1 | Gata3_ct_overlap==1)",  # Example of custom filter
    output_file = "~/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Figures/plots/Tal1.targets.neg.Tal1.Gata.pdf"
)

```

```{r}
Tal1.targets <- combined_tibble %>%
  filter(TF == "Tal1") %>%
  distinct(ensg_id) %>%
pull(ensg_id)

exp.filt <- rownames(assay.averages.rV2$RNA[rowSums(assay.averages.rV2$RNA>1.2)>0,])

non.tal1.targets <- intersect(exp.filt,rownames(assay.averages.rV2$RNA)[!rownames(assay.averages.rV2$RNA) %in% Tal1.targets])

data2plot <- log1p(assay.averages.rV2$RNA[non.tal1.targets,c("PRO1","PRO2","CO1","CO2","GA1","GA2","GA3","GA4","GA5","GA6","GL1","GL2","GL3","GL4","GL5")])

rownames(data2plot) <- sapply(rownames(data2plot),function(g){gene_id2name[[g]]})

color_scale <- colorRamp2(c(min(data2plot), max(data2plot)), c("white", "red"))
pdf(file="~/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Figures/plots/All.non.Tal1.targets.pdf", width = 8, height = 100)
Heatmap(
  data2plot,
  name = "RNA expression",
  cluster_rows = TRUE,                   # Cluster rows within each bin_category
  cluster_columns = FALSE,               # Remove column clustering
  show_row_names = TRUE,                 # Show row names (optional)
  show_column_names = TRUE,               # Show column names (optional)
  clustering_distance_rows = "euclidean",
  row_names_gp = gpar(fontsize = 4),
  clustering_method_rows = "ward.D2",
  col = color_scale
)
dev.off()
```

```{r}
tf_result <- process_TF_targets_with_TE_v2(
     Tal1.targets, 
     TF = "Tal1", 
     focus_lineage = "GA", 
     regulation_type = "Activation",
     pval_adj_threshold=0.05,
     w_mean_cons_threshold=0.5,
     zscore_threshold = 1.5,
     causality = "TF->Target",
     custom_filter = "(Gata2_ct_overlap==1 | Gata3_ct_overlap==1)"
)

plot_heatmap_with_variability_v2(
    combined_tibble = tf_result, 
    rna_data = assay.averages.rV2$RNA, 
    gene_id2name = gene_id2name,
    target_RNA_threshold = 1.2,
    output_file = "~/OneDrive - University of Helsinki/E12R1 project/Manuscript I - Regulation of Tal1 dependent rV2 lineage bifurcation/Figures/plots/Tal1.targets.GA.activation.Tal1_Gata.pdf"
)
```


# Target characterization plots {.tabset}

## Tal1 target characterization plots
```{r, fig.width=4, fig.height=6}
Tal1.targets <- filter(combined_tibble, TF=="Tal1")
plot_zscore_bar(Tal1.targets)
plot_linkage_kbp_bar(Tal1.targets)
plot_simple_variability_boxplot(Tal1.targets, assay.averages.rV2$RNA)

#plot_tal1_vs_target_scatter(gene_name2id[["Tal1"]], c("ENSMUSG00000026468", "ENSMUSG00000048540"), rna_data=assay.averages.rV2$RNA)

plot_expression_by_distance_category(Tal1.targets, rna_data=assay.rna.GABA)
plot_expression_by_distance_category(Tal1.targets, rna_data=assay.rna.GLUT)
plot_expression_by_cell_group(Tal1.targets, rna_data = log1p(assay.averages.rV2$RNA))
plot_expression_categories(Tal1.targets, rna_data = assay.averages.rV2$RNA)
```

## Gata2 target characterization plots

```{r, fig.width=4, fig.height=6}
Gata2.targets <- filter(combined_tibble, TF=="Gata2")
plot_zscore_bar(Gata2.targets)
plot_linkage_kbp_bar(Gata2.targets)
plot_simple_variability_boxplot(Gata2.targets, assay.averages.rV2$RNA)
plot_expression_by_distance_category(Gata2.targets, rna_data=assay.rna.GABA)
plot_expression_by_distance_category(Gata2.targets, rna_data=assay.rna.GLUT)
plot_expression_by_cell_group(Gata2.targets, rna_data = log1p(assay.averages.rV2$RNA))
```

## Gata3 target characterization plots

```{r, fig.width=4, fig.height=6}
Gata3.targets <- filter(combined_tibble, TF=="Gata3")
plot_zscore_bar(Gata3.targets)
plot_linkage_kbp_bar(Gata3.targets)
plot_simple_variability_boxplot(Gata3.targets, assay.averages.rV2$RNA)
plot_expression_by_distance_category(Gata3.targets, rna_data=assay.rna.GABA)
plot_expression_by_distance_category(Gata3.targets, rna_data=assay.rna.GLUT)
plot_expression_by_cell_group(Gata3.targets, rna_data = log1p(assay.averages.rV2$RNA))
```

## Vsx2 target characterization plots

```{r, fig.width=4, fig.height=6}
Vsx2.targets <- filter(combined_tibble, TF=="Vsx2")
plot_zscore_bar(Vsx2.targets)
plot_linkage_kbp_bar(Vsx2.targets)
plot_simple_variability_boxplot(Vsx2.targets, assay.averages.rV2$RNA)
plot_expression_by_distance_category(Vsx2.targets, rna_data=assay.rna.GABA)
plot_expression_by_distance_category(Vsx2.targets, rna_data=assay.rna.GLUT)
plot_expression_by_cell_group(Vsx2.targets, rna_data = log1p(assay.averages.rV2$RNA))
```


## GSEA {.tabset}

```{r Data for GSEA}
# DEG.res.GSEA.1 <- DEG.res.1$avg_log2FC
# names(DEG.res.GSEA.1) <- DEG.res.1$gene_name
# DEG.res.GSEA.1<- DEG.res.GSEA.1[DEG.res.1$p_val_adj < 0.01 & (DEG.res.1$pct.1 > 0.1 | DEG.res.1$pct.2 >0.1)]

# DEG.res.GSEA.2 <- DEG.res.2$avg_log2FC
# names(DEG.res.GSEA.2) <- DEG.res.2$gene_name
# DEG.res.GSEA.2<- DEG.res.GSEA.2[DEG.res.2$p_val_adj < 0.01 & (DEG.res.2$pct.1 > 0.1 | DEG.res.2$pct.2 >0.1)]
```

### GA/GL

```{r GA/GL GSEA plots, warning=FALSE,fig.height=5,fig.width=7}
TF_names <- c("Tal1","Gata2","Gata3","Vsx2")
set1.GSEA <- invisible(lapply(TF_names, function(target){do.GSEA(targets=targetGenes.set[[target]]$targets_gene_name,DEG.res=filter(DEG.res.1,p_val_adj < 0.01 & (pct.1 > 0.1 | pct.2 >0.1)),TF.name=target, comp.title="GA/GL DEG")}))

limits <- sapply(set1.GSEA,function(p){p$limits})
colnames(limits) <- TF_names
print(limits)

upper.limit <- set1.GSEA[[1]]$limits$upper.limit
lower.limit <- set1.GSEA[[1]]$limits$lower.limit
mid.upper.limit <- set1.GSEA[[1]]$limits$mid.upper.limit
mid.lower.limit <- set1.GSEA[[1]]$limits$mid.lower.limit

limits.to.plot <- list(upper.limit=upper.limit,lower.limit=lower.limit,mid.lower.limit=mid.lower.limit,mid.upper.limit=mid.upper.limit)

for (i in 1:length(set1.GSEA)){
 # l1 <- drawEdgeLinePlot(DEG.res=set1.GSEA[[i]], gene_name2id=gene_name2id, gene_id2name=gene_id2name, limits.to.plot=limits.to.plot, cell.groups=c("GA1_2","GL1_2"), edge_type = "leading")
  l1.GA <- plotEdgeScatterWithRawValues(DEG.res=set1.GSEA[[i]], gene_name2id=gene_name2id, gene_id2name=gene_id2name, limits.to.plot=limits.to.plot, cell.groups=c("GA1_2","GL1_2"),edge_type = "leading",log_scale = T)
  l1.GL <- plotEdgeScatterWithRawValues(DEG.res=set1.GSEA[[i]], gene_name2id=gene_name2id, gene_id2name=gene_id2name, limits.to.plot=limits.to.plot, cell.groups=c("GA1_2","GL1_2"),edge_type = "trailing",log_scale = T)
  p1 <- set1.GSEA[[i]]$p1
  p.final <- plot_grid(p1, plot_grid(l1.GA, l1.GL, ncol=2), ncol=1, rel_heights = c(1,1.5), label_size = 12)
  print(p.final)
  ggsave(plot=p.final, device = "pdf", path = plot_loc, file=paste(i,".pdf",sep=""), width = unit(8,"inch"), height = unit(7,"inch"))
}
```

```{r GA/GL GSEA for rV2 tables, warning=FALSE}
tmp.1 <- lapply(targetGenes.set,function(p){p$targets_gene_name})
names(tmp.1) <- names(targetGenes.set)
results <- fgsea(pathways = tmp.1, 
                 stats = DEG.res.GSEA.1, 
                 minSize = 5,  # Minimum size of a gene set to be considered
                 maxSize = 5000) # Maximum size of a gene set to be considered
results <- results %>% mutate(across(where(is.numeric), \(x) round(x, digits=3)))
create_dt(results)
```

# GO enrichments of leading and trailing edge target genes {.tabset}

```{r Connect Ensembl to fetch GO and set thresholds}
# Connect to the Ensembl database
ensembl <- useMart("ENSEMBL_MART_ENSEMBL", dataset = "mmusculus_gene_ensembl", host="https://nov2020.archive.ensembl.org")
gene_names_with_go <- getBM(attributes = c('external_gene_name', 'go_id'),
                       filters = 'with_go',
                       values = TRUE,
                       mart = ensembl)
# Split the data.frame into a list where each element is a character vector of GO IDs,
# and the names of the list elements are the Ensembl gene IDs
geneID2GO <- split(gene_names_with_go$go_id, gene_names_with_go$external_gene_name)
GO2geneID <- split(gene_names_with_go$external_gene_name,gene_names_with_go$go_id)
# Optionally, you can remove genes with no GO IDs if any exist
geneID2GO <- geneID2GO[sapply(geneID2GO, length) > 0]
```

```{r}
allGenes.list <- unique(dbGetQuery(con, 'SELECT DISTINCT gene_name FROM gene_metadata as gm;')[,1])
```

## Tal1 targets

### Up in GA

```{r}
geneList.fac <- as.factor(ifelse(allGenes.list %in% set1.GSEA[[1]]$edges$combined_edges$positive_leading_edge, 1,0))
names(geneList.fac) <- allGenes.list

sampleGOdata <- new("topGOdata", description = "Simple session", ontology = "BP",
allGenes = geneList.fac,
nodeSize = 5,
annot = annFUN.gene2GO, gene2GO = geneID2GO)

resultFisher <- runTest(sampleGOdata, algorithm = "classic", statistic = "fisher")
allRes <- as_tibble(GenTable(sampleGOdata, classicFisher = resultFisher, ranksOf = "classicFisher", topNodes = 50))

genes.in.classes <- sapply(allRes$GO.ID,function(GO_id){
  tmp.genes <- intersect(genesInTerm(sampleGOdata, GO_id)[[1]], sigGenes(sampleGOdata))
  return(paste(tmp.genes, collapse = ", "))
  })

allRes$genes.in.classes <- genes.in.classes

create_dt(allRes)
sigGenes(sampleGOdata)
```

### Up in GL

```{r}
geneList.fac <- as.factor(ifelse(allGenes.list %in% set1.GSEA[[1]]$edges$combined_edges$negative_trailing_edge, 1,0))
names(geneList.fac) <- allGenes.list

sampleGOdata <- new("topGOdata", description = "Simple session", ontology = "BP",
allGenes = geneList.fac,
nodeSize = 5,
annot = annFUN.gene2GO, gene2GO = geneID2GO)

resultFisher <- runTest(sampleGOdata, algorithm = "classic", statistic = "fisher")
allRes <- as_tibble(GenTable(sampleGOdata, classicFisher = resultFisher, ranksOf = "classicFisher", topNodes = 50))

genes.in.classes <- sapply(allRes$GO.ID,function(GO_id){
  tmp.genes <- intersect(genesInTerm(sampleGOdata, GO_id)[[1]], sigGenes(sampleGOdata))
  return(paste(tmp.genes, collapse = ", "))
  })

allRes$genes.in.classes <- genes.in.classes

create_dt(allRes)
sigGenes(sampleGOdata)
```

## Gata2 targets

### Up in GA

```{r}
allGenes.list <- unique(dbGetQuery(con, 'SELECT DISTINCT gene_name FROM gene_metadata as gm, links as li WHERE li.ensg_id=gm.ensg_id AND li.zscore>2;')[,1])

geneList.fac <- as.factor(ifelse(allGenes.list %in% set1.GSEA[[2]]$edges$combined_edges$positive_leading_edge, 1,0))
names(geneList.fac) <- allGenes.list

sampleGOdata <- new("topGOdata", description = "Simple session", ontology = "BP",
allGenes = geneList.fac,
nodeSize = 5,
annot = annFUN.gene2GO, gene2GO = geneID2GO)

resultFisher <- runTest(sampleGOdata, algorithm = "classic", statistic = "fisher")
allRes <- as_tibble(GenTable(sampleGOdata, classicFisher = resultFisher, ranksOf = "classicFisher", topNodes = 50))

genes.in.classes <- sapply(allRes$GO.ID,function(GO_id){
  tmp.genes <- intersect(genesInTerm(sampleGOdata, GO_id)[[1]], sigGenes(sampleGOdata))
  return(paste(tmp.genes, collapse = ", "))
  })

allRes$genes.in.classes <- genes.in.classes

create_dt(allRes)
sigGenes(sampleGOdata)
```

### Up in GL

```{r}
allGenes.list <- unique(dbGetQuery(con, 'SELECT DISTINCT gene_name FROM gene_metadata as gm, links as li WHERE li.ensg_id=gm.ensg_id AND li.zscore>2;')[,1])

geneList.fac <- as.factor(ifelse(allGenes.list %in% set1.GSEA[[2]]$edges$combined_edges$negative_trailing_edge, 1,0))
names(geneList.fac) <- allGenes.list

sampleGOdata <- new("topGOdata", description = "Simple session", ontology = "BP",
allGenes = geneList.fac,
nodeSize = 5,
annot = annFUN.gene2GO, gene2GO = geneID2GO)

resultFisher <- runTest(sampleGOdata, algorithm = "classic", statistic = "fisher")
allRes <- as_tibble(GenTable(sampleGOdata, classicFisher = resultFisher, ranksOf = "classicFisher", topNodes = 50))

genes.in.classes <- sapply(allRes$GO.ID,function(GO_id){
  tmp.genes <- intersect(genesInTerm(sampleGOdata, GO_id)[[1]], sigGenes(sampleGOdata))
  return(paste(tmp.genes, collapse = ", "))
  })

allRes$genes.in.classes <- genes.in.classes

create_dt(allRes)
sigGenes(sampleGOdata)
```

## Gata3 targets

### Up in GA

```{r}
allGenes.list <- unique(dbGetQuery(con, 'SELECT DISTINCT gene_name FROM gene_metadata as gm, links as li WHERE li.ensg_id=gm.ensg_id AND li.zscore>2;')[,1])

geneList.fac <- as.factor(ifelse(allGenes.list %in% set1.GSEA[[3]]$edges$combined_edges$positive_leading_edge, 1,0))
names(geneList.fac) <- allGenes.list

sampleGOdata <- new("topGOdata", description = "Simple session", ontology = "BP",
allGenes = geneList.fac,
nodeSize = 5,
annot = annFUN.gene2GO, gene2GO = geneID2GO)

resultFisher <- runTest(sampleGOdata, algorithm = "classic", statistic = "fisher")
allRes <- as_tibble(GenTable(sampleGOdata, classicFisher = resultFisher, ranksOf = "classicFisher", topNodes = 50))

genes.in.classes <- sapply(allRes$GO.ID,function(GO_id){
  tmp.genes <- intersect(genesInTerm(sampleGOdata, GO_id)[[1]], sigGenes(sampleGOdata))
  return(paste(tmp.genes, collapse = ", "))
  })

allRes$genes.in.classes <- genes.in.classes

create_dt(allRes)
sigGenes(sampleGOdata)
```

### Up in GL

```{r}
allGenes.list <- unique(dbGetQuery(con, 'SELECT DISTINCT gene_name FROM gene_metadata as gm, links as li WHERE li.ensg_id=gm.ensg_id AND li.zscore>2;')[,1])

geneList.fac <- as.factor(ifelse(allGenes.list %in% set1.GSEA[[3]]$edges$combined_edges$negative_trailing_edge, 1,0))
names(geneList.fac) <- allGenes.list

sampleGOdata <- new("topGOdata", description = "Simple session", ontology = "BP",
allGenes = geneList.fac,
nodeSize = 5,
annot = annFUN.gene2GO, gene2GO = geneID2GO)

resultFisher <- runTest(sampleGOdata, algorithm = "classic", statistic = "fisher")
allRes <- as_tibble(GenTable(sampleGOdata, classicFisher = resultFisher, ranksOf = "classicFisher", topNodes = 50))

genes.in.classes <- sapply(allRes$GO.ID,function(GO_id){
  tmp.genes <- intersect(genesInTerm(sampleGOdata, GO_id)[[1]], sigGenes(sampleGOdata))
  return(paste(tmp.genes, collapse = ", "))
  })

allRes$genes.in.classes <- genes.in.classes

create_dt(allRes)
sigGenes(sampleGOdata)
```

## Vsx2 targets

### Up in GA

```{r}
allGenes.list <- unique(dbGetQuery(con, 'SELECT DISTINCT gene_name FROM gene_metadata as gm, links as li WHERE li.ensg_id=gm.ensg_id AND li.zscore>2;')[,1])

geneList.fac <- as.factor(ifelse(allGenes.list %in% set1.GSEA[[4]]$edges$combined_edges$positive_leading_edge, 1,0))
names(geneList.fac) <- allGenes.list

sampleGOdata <- new("topGOdata", description = "Simple session", ontology = "BP",
allGenes = geneList.fac,
nodeSize = 5,
annot = annFUN.gene2GO, gene2GO = geneID2GO)

resultFisher <- runTest(sampleGOdata, algorithm = "classic", statistic = "fisher")
allRes <- as_tibble(GenTable(sampleGOdata, classicFisher = resultFisher, ranksOf = "classicFisher", topNodes = 50))

genes.in.classes <- sapply(allRes$GO.ID,function(GO_id){
  tmp.genes <- intersect(genesInTerm(sampleGOdata, GO_id)[[1]], sigGenes(sampleGOdata))
  return(paste(tmp.genes, collapse = ", "))
  })

allRes$genes.in.classes <- genes.in.classes

create_dt(allRes)
sigGenes(sampleGOdata)
```

### Up in GL

```{r}
allGenes.list <- unique(dbGetQuery(con, 'SELECT DISTINCT gene_name FROM gene_metadata as gm, links as li WHERE li.ensg_id=gm.ensg_id AND li.zscore>2;')[,1])

geneList.fac <- as.factor(ifelse(allGenes.list %in% set1.GSEA[[4]]$edges$combined_edges$negative_trailing_edge, 1,0))
names(geneList.fac) <- allGenes.list

sampleGOdata <- new("topGOdata", description = "Simple session", ontology = "BP",
allGenes = geneList.fac,
nodeSize = 5,
annot = annFUN.gene2GO, gene2GO = geneID2GO)

resultFisher <- runTest(sampleGOdata, algorithm = "classic", statistic = "fisher")
allRes <- as_tibble(GenTable(sampleGOdata, classicFisher = resultFisher, ranksOf = "classicFisher", topNodes = 50))

genes.in.classes <- sapply(allRes$GO.ID,function(GO_id){
  tmp.genes <- intersect(genesInTerm(sampleGOdata, GO_id)[[1]], sigGenes(sampleGOdata))
  return(paste(tmp.genes, collapse = ", "))
  })

allRes$genes.in.classes <- genes.in.classes

create_dt(allRes)
sigGenes(sampleGOdata)
```

```{r, eval=FALSE}
# Generate a series of threshold values
thresholds <- seq(0, 25, by = 1)

# Create a new data frame with an added threshold condition column
thresholded_data <- do.call(rbind, lapply(thresholds, function(b.thr) {
    combined_tibble %>%
        filter(TF == "Gata2" & TF.cor.PRO.GA > 0.65) %>%
        mutate(threshold = b.thr, 
               meets_threshold = sum_GA1_2_bound > b.thr)
}))

# Filter the combined data to only include rows that meet the threshold condition
thresholded_data <- thresholded_data %>% filter(meets_threshold)

# Create the faceted plot
plot <- ggplot(thresholded_data, aes(x = TF.cor.PRO.GA, y = PRO1_2_GA_increase)) +
    geom_point() +
    geom_smooth(method = "lm") +
    theme_minimal() +
    stat_cor(label.x = 0.65, label.y = 1.3, aes(label = after_stat(r.label)), method = "spearman") +
    facet_wrap(~ threshold, scales = "free")

# Print the plot
print(plot)
```

```{r, eval=FALSE}
# Generate a series of threshold values
thresholds <- seq(0, 25, by = 1)

# Create a new data frame with an added threshold condition column
thresholded_data <- do.call(rbind, lapply(thresholds, function(b.thr) {
    combined_tibble %>%
        filter(TF == "Gata3" & TF.cor.PRO.GA > 0.65) %>%
        mutate(threshold = b.thr, 
               meets_threshold = sum_GA1_2_bound > b.thr)
}))

# Filter the combined data to only include rows that meet the threshold condition
thresholded_data <- thresholded_data %>% filter(meets_threshold)

# Create the faceted plot
plot <- ggplot(thresholded_data, aes(x = TF.cor.PRO.GA, y = PRO1_2_GA_increase)) +
    geom_point() +
    geom_smooth(method = "lm") +
    theme_minimal() +
    stat_cor(label.x = 0.65, label.y = 1.3, aes(label = after_stat(r.label)), method = "spearman") +
    facet_wrap(~ threshold, scales = "free")

# Print the plot
print(plot)
```
