---
title: "R Notebook of TOBIAS table operations"
output: html_document
---

```{r Packages, message=FALSE}
library(rtracklayer)
library(qs)
library(Seurat)
library(Signac)
library(tidyverse)
library(GenomicRanges)
library(parallel)
library(dbplyr)
library(DBI)
```

```{r Additional functions}
message_parallel <- function(...){
  system(sprintf('echo "\n%s\n"', paste0(..., collapse="")))
}
```

```{r Reading Seurat dataobject}
dataset <- qread("../scATAC_data/nmm_rV2_subset_relabeled_031023_links.qs", nthreads = 10)
DefaultAssay(dataset) <- "RNA"

dataset$rv2.lineage_re <- case_when(
  dataset$rv2.lineage %in% "PRO1" ~ "PRO1_2",
  dataset$rv2.lineage %in% "PRO2" ~ "PRO1_2",
  dataset$rv2.lineage %in% "GA1" ~ "GA1_2",
  dataset$rv2.lineage %in% "GA2" ~ "GA1_2",
  dataset$rv2.lineage %in% "GA3" ~ "GA3_4",
  dataset$rv2.lineage %in% "GA4" ~ "GA3_4",
  dataset$rv2.lineage %in% "GA5" ~ "GA5_6",
  dataset$rv2.lineage %in% "GA6" ~ "GA5_6",
  dataset$rv2.lineage %in% "CO1" ~ "CO1_2",
  dataset$rv2.lineage %in% "CO2" ~ "CO1_2",
  dataset$rv2.lineage %in% "GL1" ~ "GL1_2",
  dataset$rv2.lineage %in% "GL2" ~ "GL1_2",
  dataset$rv2.lineage %in% "GL3" ~ "GL3_4",
  dataset$rv2.lineage %in% "GL4" ~ "GL3_4",
  dataset$rv2.lineage %in% "GL5" ~ "GL5",
)

DefaultAssay(dataset) <- "peaks"
all.features <- StringToGRanges(rownames(dataset))
```

```{r Reading Tobias results and H12 metadata, eval=FALSE}
#rV2.groups.tobias.h12.gr.dr <- qread("../analysis/rV2.groups.tobias.h12.gr.dr.regrouped.qs", nthreads = 8)

rV2.groups.tobias.h12.gr.dr <- get_BINDetect_snakemake_results_gr(res_path="/Volumes/VarastoNAS/UniversityData/TOBIAS_090224/TFBS/",parallel=T, mc.cores=2, HOCOMOCO=12)

#H12.metadata <- qread(file="../analysis/H12_metadata_mod.qs")
#H12.metadata$TF.rep <- paste(H12.metadata$name,"_",H12.metadata$name,sep="")
#write_csv(H12.metadata, col_names = TRUE, file = "../metadata/H12.data.csv")
H12.metadata <- read_delim("../metadata/H12.data_INSM1_added.csv", col_names = TRUE, delim = ";")
```

```{r TRansform tobias gr to tb, eval=FALSE}
tobias.tb.list <- lapply(rV2.groups.tobias.h12.gr.dr, as_tibble)
tobias.tb <- do.call(rbind,tobias.tb.list)
qsave(tobias.tb, file = "../analysis/Tobias.rv2.h12.dr.tb.qs",nthreads = 10)
```

```{r Finding TFBS to feature matches, eval=FALSE}
TFBS.pos.tb <- tobias.tb %>% select(seqnames, start, end, TFBS_name)
TFBS.to.features <- mergeByOverlaps(query = makeGRangesFromDataFrame(TFBS.pos.tb, keep.extra.columns = T), subject = all.features,type="within")

TFBS.to.features.tb <- cbind(as_tibble(TFBS.to.features$`makeGRangesFromDataFrame(TFBS.pos.tb, keep.extra.columns = T)`), features=GRangesToString(TFBS.to.features$all.features))
                             
tobias.tb.2 <- left_join(tobias.tb, select(TFBS.to.features.tb, seqnames, start, end, TFBS_name, features))
rm(TFBS.to.features.tb)
rm(tobias.tb)
rm(rV2.groups.tobias.h12.gr.dr)
gc()

qsave(tobias.tb.2, file = "../analysis/Tobias.rv2.h12.dr.tb.2.qs",nthreads = 10)
```

```{r Connection to SQLite}
con <- DBI::dbConnect(RSQLite::SQLite(), dbname = "~/Workspace/TOBIAS.dr.h12_030324.sqlite")

```

```{r Write gene.metadata into SQlite}
gene.metadata <- dataset[['RNA']][[]] %>% rownames_to_column("ensg_id") %>% rename("feature_symbol"="gene_name")
dbWriteTable(con.obj, "gene_metadata", gene.metadata, overwrite=TRUE)
dbExecute(con.obj, 'CREATE INDEX ensg_id3 ON gene_metadata (ensg_id);')
dbExecute(con.obj, 'CREATE INDEX gene_name3 ON gene_metadata (gene_name);')
```

```{r Write features table into SQLite, eval=FALSE}
features.tb <- as_tibble(all.features)
features.tb$feature <- paste(features.tb$seqnames,"-",features.tb$start,"-",features.tb$end,sep="")

dbWriteTable(con.obj, "feature", features.tb, overwrite=TRUE)
dbExecute(con.obj, 'CREATE INDEX features2 ON feature (feature);')
dbExecute(con.obj, 'CREATE INDEX seqnames2 ON feature (seqnames);')
dbExecute(con.obj, 'CREATE INDEX start2 ON feature (start);')
dbExecute(con.obj, 'CREATE INDEX end2 ON feature (end);')
```

```{r Write TOBIAS data into SQlite, eval=FALSE}
dbWriteTable(con, "tobias", tobias.tb.2, overwrite=TRUE)
dbListTables(con)
rm(tobias.tb.2)
gc()
```

```{r Build indeces, eval=FALSE}
dbExecute(con, 'CREATE INDEX seqname 
ON tobias (seqnames);')

dbExecute(con, 'CREATE INDEX position 
ON tobias (seqnames, start, end);')

dbExecute(con, 'CREATE INDEX TFBS_name 
ON tobias (TFBS_name);')

dbExecute(con, 'CREATE INDEX features 
ON tobias (features);')
```

```{r}
tobias.tm <- dbGetQuery(con, 'SELECT TFBS_name FROM tobias')

tobias.tm <- left_join(tobias.tm, select(H12.metadata, TF.rep, masterlist_info.tf,ensg_id), by=c("TFBS_name"="TF.rep")) %>% distinct()
colnames(tobias.tm) <- c("TFBS_name","masterlist_info_tf","ensg_id")

dbWriteTable(con, "gene_name_tmp", tobias.tm)
dbExecute(con, 'CREATE INDEX TFBS_name_2 
ON gene_name_tmp (TFBS_name);')

dbExecute(con, 'ALTER TABLE tobias ADD COLUMN TF_gene_name TEXT;')
dbExecute(con, 'ALTER TABLE tobias ADD COLUMN ensg_id TEXT;')

dbExecute(con, '
UPDATE tobias
SET TF_gene_name = (
    SELECT masterlist_info_tf
    FROM gene_name_tmp
    WHERE gene_name_tmp.TFBS_name = tobias.TFBS_name
),
ensg_id = (
    SELECT ensg_id
    FROM gene_name_tmp
    WHERE gene_name_tmp.TFBS_name = tobias.TFBS_name
);
')

dbExecute(con, 'DROP TABLE gene_name_tmp;')

dbExecute(con, 'CREATE INDEX TF_gene_name ON tobias (TF_gene_name);')
dbExecute(con, 'CREATE INDEX ensg_id ON tobias (ensg_id);')
gc()
```

```{r Reading conservation data}
cons.gr <- import.bw(con="../metadata/mm10.60way.phastCons.bw",as="GRanges")
```

```{r}
TFBS.loc <- dbGetQuery(con, 'SELECT seqnames, start, end FROM tobias WHERE seqnames IN ("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chrX","chrY");')
```

```{r }
uniq.seq.names <- unique(pull(TFBS.loc,seqnames))

TFBS.loc.cons.list <- mclapply(uniq.seq.names, function(seq.name){
  message_parallel(paste("Processing",seq.name,sep=" "))
  gc()
  TFBS.tmp <- makeGRangesFromDataFrame(filter(TFBS.loc, seqnames==seq.name))
  cons.subset <- cons.gr[seqnames(cons.gr)==seq.name]
  TFBS.cons <- mergeByOverlaps(query = cons.subset, subject = TFBS.tmp, type="within")
  mcols(TFBS.cons$TFBS.tmp)$score <- TFBS.cons$score
  TFBS.cons.means <- as_tibble(TFBS.cons$TFBS.tmp %>% plyranges::group_by(start,end) %>% summarize(mean.cons=mean(score)))
  TFBS.cons.means$seqnames <-seq.name
  return(TFBS.cons.means)
}, mc.cores = 1)


TFBS.loc.cons.tb <- do.call(rbind,TFBS.loc.cons.list)
colnames(TFBS.loc.cons.tb) <- c("start","end","mean_cons","seqnames")
tobias.cons.tmp <- left_join(TFBS.loc, TFBS.loc.cons.tb, by=c("seqnames","start","end"))

dbWriteTable(con, "TFBS_pos_cons_tmp", tobias.cons.tmp, overwrite=TRUE)
dbExecute(con, 'CREATE INDEX TFBS_loc_2
ON TFBS_pos_cons_tmp (seqnames, start, end);')

dbExecute(con, 'ALTER TABLE tobias ADD COLUMN mean_cons REAL;')

dbExecute(con, '
UPDATE tobias
SET mean_cons = (
    SELECT mean_cons
    FROM TFBS_pos_cons_tmp
    WHERE TFBS_pos_cons_tmp.seqnames = tobias.seqnames AND TFBS_pos_cons_tmp.start = tobias.start AND TFBS_pos_cons_tmp.end = tobias.end
)')


# Not all TFBS get conservation score, is this becuase of the conservation source data??

dbExecute(con, 'DROP TABLE TFBS_pos_cons_tmp;')
```

# Calculate expression and accessibility information into the database

```{r Calculate and upload average gene expression data per cell group into separate table}
ensg_ids <- dbGetQuery(con, 'SELECT DISTINCT ensg_id FROM tobias;')[,1]

TF.expression.data.avg2 <- AverageExpression(dataset, assays = "RNA", features = ensg_ids, group.by = "rv2.lineage_re", layer = "data")[[1]]

dbWriteTable(con, "exp", as.data.frame(TF.expression.data.avg2) %>% rownames_to_column("ensg_id") %>% as_tibble(), overwrite=TRUE)

dbExecute(con, 'CREATE INDEX ensg_ids_2 ON exp (ensg_id);')
```

```{r Calculate and upload z-score normalized gene expression averages per TF gene per cell group into separate table}
ensg_ids <- dbGetQuery(con, 'SELECT DISTINCT ensg_id FROM tobias;')[,1]

DefaultAssay(dataset) <- "RNA"
TF.expression.data <- FetchData(object=dataset, assays = "RNA", vars = c(ensg_ids,"rv2.lineage_re"))
TF.expression.data.mat <- as.matrix(expm1(as_tibble(TF.expression.data) %>% select(!starts_with("rv"))))
TF.expression.data.mat <- as_tibble(t(scale(t(TF.expression.data.mat))))
TF.expression.data.mat$rv2.lineage_re <- TF.expression.data$rv2.lineage_re
TF.expression.data.mat <- pivot_longer(TF.expression.data.mat, cols = !starts_with("rv"))

TF.expr.scaled <- TF.expression.data.mat %>% group_by(rv2.lineage_re, name) %>% summarise(avg.exp=mean(value,na.rm=T)) %>% pivot_wider(values_from = avg.exp, names_from = rv2.lineage_re)

names(TF.expr.scaled)[1] <- "ensg_id"

dbWriteTable(con, "exp_scaled", TF.expr.scaled, overwrite=TRUE)

dbExecute(con, 'CREATE INDEX ensg_ids_3 ON exp_scaled (ensg_id);')
```

```{r Calculate and upload feature accessibility data per feature per cell group into separate table }
features <- dbGetQuery(con, 'SELECT DISTINCT features FROM tobias;')[,1]

acc.avg.data <- AverageExpression(dataset, assays = "peaks", features = features, group.by = "rv2.lineage_re")[[1]]

dbWriteTable(con, "acc", as.data.frame(acc.avg.data) %>% rownames_to_column("features") %>% as_tibble(), overwrite=TRUE)

dbExecute(con, 'CREATE INDEX features_2 ON acc (features);')
```

```{r Finding consequent motif areas}
tobias.table <- tbl(con, "tobias")
exp.table <- tbl(con, "exp")
acc.table <- tbl(con, "acc")
table.tmp.1 <- tobias.table %>% left_join(exp.table) %>% left_join(acc.table, by=c("features"="features")) %>% dplyr::filter(mean_cons>0.5)
table.tmp.2 <- table.tmp.1 %>% collect()

acc.thr <- quantile(acc.table %>% dplyr::select(!starts_with("features")) %>% pull(),.25)
mean_cons_thr<-0.5

table.tmp.2 <- table.tmp.2 %>% dplyr::filter((abs(PRO1_2.x)>1.2 | abs(CO1_2.x)>1.2 | abs(GA1_2.x)>1.2 | abs(GL1_2.x)>1.2) & (abs(PRO1_2.y)>acc.thr | abs(CO1_2.y)>acc.thr | abs(GA1_2.y)>acc.thr | abs(GL1_2.y)>acc.thr) & (PRO1_2_bound==1 | CO1_2_bound==1 | GA1_2_bound==1 | GL1_2_bound==1) & mean_cons>mean_cons_thr) %>% arrange(start)

grB <- makeGRangesFromDataFrame(dplyr::select(table.tmp.2, seqnames, start, end, TFBS_name), keep.extra.columns = T)

reduced_grB_with_map <- reduce(grB, min.gapwidth=0L,with.revmap = TRUE)


# Step 2: Concatenate metadata based on the revmap information
concatenated_metadata <- mclapply(reduced_grB_with_map$revmap, function(indices) {
  if (length(indices) > 0) {
    meta_values <- mcols(grB)[indices, ]
    paste(unique(na.omit(meta_values)), collapse = "; ")
  } else {
    ""
  }
}, mc.cores = 10)

mcols(reduced_grB_with_map)$concatenated_metadata <- concatenated_metadata

plyranges::filter_by_overlaps(reduced_grB_with_map,StringToGRanges("chr4-114432894-115134633"))
mcols(plyranges::filter_by_overlaps(reduced_grB_with_map,StringToGRanges("chr4-114432894-115134633")))$concatenated_metadata

saveRDS(reduced_grB_with_map,"../analysis/TFBS_reduced_positions_120324.Rds")
```

