---
title: "R Notebook of two ghosts analysis"
output: html_notebook
---

```{r Packages, include=FALSE, echo=FALSE}
library(tidyverse)
library(GenomicRanges)
library(qs)
library(Signac)
library(Seurat)
library(circlize)
library(valr)
library(ggVennDiagram)
source("../../AuxCode/AuxFunctions.R")
```

```{r Define TADs}
Tal1.TAD <- StringToGRanges(construct_range(4, 114300000, 115750000, 0))
Tal1.feat.6 <- StringToGRanges("chr4-115056171-115058012")
Tal1.feat.11 <- StringToGRanges("chr4-115078530-115080241")

Gata2.TAD <- StringToGRanges(construct_range(6, 88100000, 88450000, 0))
```

```{r load rV2 dataset}
dataset <- qread("../scATAC_data/nmm_rV2_subset_relabeled_031023_links.qs",nthreads = 6)
DefaultAssay(dataset) <- "peaks"
all.features.gr <- StringToGRanges(rownames(dataset))
write_tsv(gr_to_bed(all.features.gr), col_names = FALSE, file="../analysis/all.features.bed")
```

# Ghosts I
```{r Read Gata2 EChO data}
Gata2.EChO.mat <- as.matrix(read_tsv("../analysis/EChO/Gata2_R1_ab_matrix.EChO.matrix", col_names = FALSE))
Gata2.span <- read_tsv("../analysis/EChO/Gata2_R1_ab_matrix.spans.txt", col_names = c("line","span","s","fit","se"))
Gata2.foci <- read_tsv("../analysis/EChO/Gata2_R1_ab_foci.EChO.bed", col_names = c("chr","start","end","frag.size.mean","frag.size.var","span"))
```

```{r Read TAl1 EChO data}
Tal1.EChO.mat <- as.matrix(read_tsv("../analysis/EChO/Tal1_R1_757_matrix.EChO.matrix", col_names = FALSE))
Tal1.span <- read_tsv("../analysis/EChO/Tal1_R1_757_matrix.spans.txt", col_names = c("line","span","s","fit","se"))
Tal1.foci <- read_tsv("../analysis/EChO/Tal1_R1_757_foci.EChO.bed", col_names = c("chr","start","end","frag.size.mean","frag.size.var","span"))
```

```{r Find Gata2 echos}
# gata2.echos <- findEChO(Gata2.EChO.mat, span=Gata2.span, foci = Gata2.foci)
# gata2.echos.gr <- StringToGRanges(gata2.echos$coordinates)
# searc.Gata2.fimo.gr <- makeGRangesFromDataFrame(filter(Seacr.Gata2, str_detect(motif_id,"GATA2.*")) %>% select(c("sequence_name","start","stop")),seqnames.field = "sequence_name", na.rm=T)
# gata2.over1 <- reduce(gata2.echos.gr[gata2.echos.gr %over% searc.Gata2.fimo.gr])
```

```{r Find Tal1 echos}
# tal1.echos <- findEChO(Tal1.EChO.mat, span=Tal1.span, foci = Tal1.foci)
# tal1.echos.gr <- StringToGRanges(tal1.echos$coordinates)
# searc.Tal1.fimo.gr <- makeGRangesFromDataFrame(filter(Seacr.Tal1, str_detect(motif_id,"TAL1.*")) %>% select(c("sequence_name","start","stop")),seqnames.field = "sequence_name", na.rm=T)
# tal1.over1 <- reduce(tal1.echos.gr[tal1.echos.gr %over% searc.Tal1.fimo.gr])
```

```{r Process Tal1 foci locations from EChOs}
Tal1.foci.gr <- makeGRangesFromDataFrame(data.frame(chr=Tal1.foci$chr, start=Tal1.foci$start, end=Tal1.foci$end))
#Tal1.foci$dist <- elementMetadata(distanceToNearest(x=Tal1.foci.gr, subject=searc.Tal1.fimo.gr))$distance
Tal1.foci$direct <- as.factor(ifelse(Tal1.foci$frag.size.mean<=120,1,0))
```

```{r Process Gata2 foci locations from EChOs}
Gata2.foci.gr <- makeGRangesFromDataFrame(data.frame(chr=Gata2.foci$chr, start=Gata2.foci$start, end=Gata2.foci$end))
Gata2.foci$direct <- as.factor(ifelse(Gata2.foci$frag.size.mean<=120,1,0))
```

# Ghosts II

```{r Read H12 TOBIAS data}
rV2.groups.tobias.h12.gr <- qread("../analysis/rV2.groups.tobias.H12.gr.qs",nthreads = 6)
```

```{r Gata2 in groups}
GA1.Gata2.gr <- ConstructBed_TobiasGr(gr=rV2.groups.tobias.h12.gr, group="GA1",TF="Gata2", gr.only = T, file = NULL)
GL1.Gata2.gr <- ConstructBed_TobiasGr(gr=rV2.groups.tobias.h12.gr, group="GL1",TF="Gata2", gr.only = T, file = NULL)
CO2.Gata2.gr <- ConstructBed_TobiasGr(gr=rV2.groups.tobias.h12.gr, group="CO2",TF="Gata2", gr.only = T, file = NULL)
```

```{r Tal1 in groups}
GA1.Tal1.gr <- ConstructBed_TobiasGr(gr=rV2.groups.tobias.h12.gr, group="GA1",TF="Tal1", gr.only = T, file = NULL)
CO2.Tal1.gr <- ConstructBed_TobiasGr(gr=rV2.groups.tobias.h12.gr, group="CO2",TF="Tal1", gr.only = T, file = NULL)
GL1.Tal1.gr <- ConstructBed_TobiasGr(gr=rV2.groups.tobias.h12.gr, group="GL1",TF="Tal1", gr.only = T, file = NULL)
```

# Distances between ghosts +- 200bp

```{r Gata2 distances between ghosts}
Gata2.dir.echos.gr <- filter(Gata2.foci, direct==1) %>% makeGRangesFromDataFrame()

GA1.Gata2.ghost.distances <- distanceToNearest(x=Gata2.dir.echos.gr, subject = GA1.Gata2.gr)
GA1.Gata2.double.manifest <- GA1.Gata2.gr[subjectHits(GA1.Gata2.ghost.distances[elementMetadata(GA1.Gata2.ghost.distances)$distance<200])]

GL1.Gata2.ghost.distances <- distanceToNearest(x=Gata2.dir.echos.gr, subject = GL1.Gata2.gr)
GL1.Gata2.double.manifest <- GL1.Gata2.gr[subjectHits(GL1.Gata2.ghost.distances[elementMetadata(GL1.Gata2.ghost.distances)$distance<200])]

CO2.Gata2.ghost.distances <- distanceToNearest(x=Gata2.dir.echos.gr, subject = CO2.Gata2.gr)
CO2.Gata2.double.manifest <- CO2.Gata2.gr[subjectHits(CO2.Gata2.ghost.distances[elementMetadata(CO2.Gata2.ghost.distances)$distance<200])]
```

```{r Tables of Gata2 nearest genes}
create_dt(as_tibble(GA1.Gata2.double.manifest))
create_dt(as_tibble(GL1.Gata2.double.manifest))
create_dt(as_tibble(CO2.Gata2.double.manifest))
```

```{r Venn diagram of nearest genes for Gata2 ghost double manifestations}
ggVennDiagram(list(GA1=pull(as_tibble(GA1.Gata2.double.manifest),gene_name),GL1=pull(as_tibble(GL1.Gata2.double.manifest),gene_name),CO2=pull(as_tibble(CO2.Gata2.double.manifest),gene_name)))
```

```{r Tal1 distances between ghosts}
Tal1.dir.echos.gr <- filter(Tal1.foci, direct==1) %>% makeGRangesFromDataFrame()

GA1.Tal1.ghost.distances <- distanceToNearest(x=Tal1.dir.echos.gr, subject = GA1.Tal1.gr)
GA1.Tal1.double.manifest <- GA1.Tal1.gr[subjectHits(GA1.Tal1.ghost.distances[elementMetadata(GA1.Tal1.ghost.distances)$distance<200])]

GL1.Tal1.ghost.distances <- distanceToNearest(x=Tal1.dir.echos.gr, subject = GL1.Tal1.gr)
GL1.Tal1.double.manifest <- GL1.Tal1.gr[subjectHits(GL1.Tal1.ghost.distances[elementMetadata(GL1.Tal1.ghost.distances)$distance<200])]

CO2.Tal1.ghost.distances <- distanceToNearest(x=Tal1.dir.echos.gr, subject = CO2.Tal1.gr)
CO2.Tal1.double.manifest <- CO2.Tal1.gr[subjectHits(CO2.Tal1.ghost.distances[elementMetadata(CO2.Tal1.ghost.distances)$distance<200])]
```

```{r Tables of Tal1 nearest genes}
create_dt(as_tibble(GA1.Tal1.double.manifest))
create_dt(as_tibble(GL1.Tal1.double.manifest))
create_dt(as_tibble(CO2.Tal1.double.manifest))
```

```{r Venn diagram of nearest genes for Tal1 ghost double manifestations}
ggVennDiagram(list(GA1=pull(as_tibble(GA1.Tal1.double.manifest),gene_name),GL1=pull(as_tibble(GL1.Gata2.double.manifest),gene_name),CO2=pull(as_tibble(CO2.Gata2.double.manifest),gene_name)))
```


```{r Echos in Tal1 pos control areas}
Tal1.in.Tal1.EChOs.pos.controls <- c(Tal1.dir.echos.gr[Tal1.dir.echos.gr %over% Tal1.feat.6], Tal1.dir.echos.gr[Tal1.dir.echos.gr %over% Tal1.feat.11])
write_tsv(gr_to_bed(Tal1.in.Tal1.EChOs.pos.controls), col_names = FALSE, file = "../analysis/Tal1.in.Tal1.EChOs.pos.controls.bed")

Gata2.in.Tal1.EChOs.pos.controls <- c(Gata2.dir.echos.gr[Gata2.dir.echos.gr %over% Tal1.feat.6], Gata2.dir.echos.gr[Gata2.dir.echos.gr %over% Tal1.feat.11])
write_tsv(gr_to_bed(Gata2.dir.echos.gr), col_names = FALSE, file = "../analysis/Gata2.in.Tal1.EChOs.pos.controls.bed")
```

```{r TOBIAS GA1 in Tal1 pos control areas}
Tal1.in.Tal1.TOBIAS.pos.controls <- c(GA1.Tal1.gr[GA1.Tal1.gr %over% Tal1.feat.6], GA1.Tal1.gr[GA1.Tal1.gr %over% Tal1.feat.11])
write_tsv(gr_to_bed(Tal1.in.Tal1.TOBIAS.pos.controls), col_names = FALSE, file = "../analysis/Tal1.in.Tal1.GA1.TOBIAS.pos.controls.bed")

Gata2.in.Tal1.TOBIAS.pos.controls <- c(GA1.Gata2.gr[GA1.Gata2.gr %over% Tal1.feat.6], GA1.Gata2.gr[GA1.Gata2.gr %over% Tal1.feat.11])
write_tsv(gr_to_bed(Gata2.in.Tal1.TOBIAS.pos.controls), col_names = FALSE, file = "../analysis/Gata2.in.Tal1.GA1.TOBIAS.pos.controls.bed")
```

```{r TOBIAS CO2 in Tal1 pos control areas}
Tal1.in.Tal1.TOBIAS.pos.controls <- c(CO2.Tal1.gr[CO2.Tal1.gr %over% Tal1.feat.6], CO2.Tal1.gr[CO2.Tal1.gr %over% Tal1.feat.11])
write_tsv(gr_to_bed(Tal1.in.Tal1.TOBIAS.pos.controls), col_names = FALSE, file = "../analysis/Tal1.in.Tal1.CO2.TOBIAS.pos.controls.bed")

Gata2.in.Tal1.TOBIAS.pos.controls <- c(CO2.Gata2.gr[CO2.Gata2.gr %over% Tal1.feat.6], CO2.Gata2.gr[CO2.Gata2.gr %over% Tal1.feat.11])
write_tsv(gr_to_bed(Gata2.in.Tal1.TOBIAS.pos.controls), col_names = FALSE, file = "../analysis/Gata2.in.Tal1.CO2.TOBIAS.pos.controls.bed")
```

```{r TOBIAS GL1 in Tal1 pos control areas}
Tal1.in.Tal1.TOBIAS.pos.controls <- c(GL1.Tal1.gr[GL1.Tal1.gr %over% Tal1.feat.6], GL1.Tal1.gr[GL1.Tal1.gr %over% Tal1.feat.11])
write_tsv(gr_to_bed(Tal1.in.Tal1.TOBIAS.pos.controls), col_names = FALSE, file = "../analysis/Tal1.in.Tal1.GL1.TOBIAS.pos.controls.bed")

Gata2.in.Tal1.TOBIAS.pos.controls <- c(GL1.Gata2.gr[GL1.Gata2.gr %over% Tal1.feat.6], GL1.Gata2.gr[GL1.Gata2.gr %over% Tal1.feat.11])
write_tsv(gr_to_bed(Gata2.in.Tal1.TOBIAS.pos.controls), col_names = FALSE, file = "../analysis/Gata2.in.Tal1.GL1.TOBIAS.pos.controls.bed")
```


```{r Gata2 TOBIAS data in Tal1 feat 11}
Gata2.TOB.gr <- rV2.groups.tobias.h12.gr$GATA2.H12CORE.1.P.B_GATA2.H12CORE.1.P.B

Gata2.Tal1.feat.11.gr <- Gata2.TOB.gr[Gata2.TOB.gr %over% Tal1.feat.11]

```



# Junk below


```{r Checking more accurate positional mapping}
direct.Tal1 <- Tal1.foci[Tal1.span[tal1.echos$EChO.true.i,]$line,]
direct.Tal1.foci.gr <- makeGRangesFromDataFrame(data.frame(chr=pull(direct.Tal1,"chr"),start=pull(direct.Tal1,"foci"), end=pull(direct.Tal1,"foci")))

sum(countOverlaps(direct.Tal1.foci.gr, searc.Tal1.fimo.gr, minoverlap = 0L, maxgap = 100, type="any"))

tmp1 <- sapply(0:10000,function(gap){sum(countOverlaps(direct.Tal1.foci.gr, searc.Tal1.fimo.gr, minoverlap = 0L, maxgap = gap, type="any"))})
```



```{r}
ggscatter(data=Tal1.foci, y="frag.size.mean", x="dist", color="direct",ylab="avg. frag. size",xlab="distance to nearest motif",palette = c("black", "red"))
```


```{r Combined distances ECho TOBIAS}
#  Combined distances ECho TOBIAS  <200, note FIMO conflicts TOBIAS
d1 <- distanceToNearest(makeGRangesFromDataFrame(filter(Tal1.foci,frag.size.mean<120)), CO2.Tal1.gr)
CO2.res <- CO2.Tal1.gr[subjectHits(d1[elementMetadata(d1)$distance<200])]

d2 <- distanceToNearest(makeGRangesFromDataFrame(filter(Tal1.foci,frag.size.mean<120)), GA1.Tal1.gr)
GA1.res <- GA1.Tal1.gr[subjectHits(d2[elementMetadata(d2)$distance<200])]

d3 <- distanceToNearest(makeGRangesFromDataFrame(filter(Tal1.foci,frag.size.mean<120)), GL1.Tal1.gr)
GL1.res <- GL1.Tal1.gr[subjectHits(d3[elementMetadata(d3)$distance<200])]
```

